<!DOCTYPE html>
<html>
  <head>
    <title>Genomic Features In Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Genomic Features In Bioconductor
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Bioconductor_Introduction/" class="uri">http://rockefelleruniversity.github.io/Bioconductor_Introduction/</a>

---





# Genomic Features

Genomic Features are often stored as GTF or GFF files.

&lt;div align="center"&gt;
&lt;img src="imgs/gff.png" alt="igv" height="400" width="600"&gt;
&lt;/div&gt;

---
# Genes as Genomic Features

In particular, we often find gene models as GFF or GTF files.

This importantly allows us to associate.

* Exons with transcripts
* Transcripts with genes.

With this mapping we can perform complex operations such as summarising genome wide RNAseq signal to genes'/transcripts' expressions.  

---
# IGV and GFF/GTF.

We use GFF and GTF in IGV to display gene models. 



![](imgs/fullGenes.png)

---
# A single gene model in GTF format.

GTF is perhaps the most common format for gene models. 

In this example we have a GTF file containing a single gene model for the **Xkr4** gene.

This can be found in **Data/Xkr4.gtf**

![](imgs/gtfInTerminal.png)
---
# A single gene model in IGV.

IGV parses information from the GTF file to construct the gene model we see containing untranslated regions and directionality of expression.

![](imgs/gtfInIGV.png)

---
# Genomic Features in Bioconductor .

In Bioconductor, gene models are prepackaged for us in the **TxDb** packages.

Format is TxDb. **species** . **source** . **major version** . **table**

Homo Sapiens gene build from UCSC's version hg19 known gene table -- TxDb.Hsapiens.UCSC.hg19.knownGene

&lt;div align="center"&gt;
&lt;img src="imgs/txdb.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;


---
# Genomic Features in Bioconductor.

We can however make use of GTF/GFF files of our own gene models outside of Bioconductor pre-built TxDB packages.

The **GenomicFeatures** package has functions to import a GTF/GFF into a Bioconductor **TxDb** object as well as to interact with **TxDb** objects.

- [**GenomicFeatures**](https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html) -- Import to and handling of **TxDB** objects in R.

---
# Genomic Features in Bioconductor.

The first package we will look at is the **TxDB** package - **TxDb.Hsapiens.UCSC.hg19.knownGene**.

Remember we can install Bioconductor packages (fairly) easily  using Bioconductor package pages' provided commands. 


```r
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
```

---
# Genomic Intervals in Bioconductor.

Now we have the package installed, we can load the library




```r
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

---
# TxDb

Once we have loaded the **TxDb.Hsapiens.UCSC.hg19.knownGene** library, we will have a new **TxDb** object available to us called **TxDb.Hsapiens.UCSC.hg19.knownGene** 


```r
class(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

```
## [1] "TxDb"
## attr(,"package")
## [1] "GenomicFeatures"
```
---
# TxDb

When this variable name is typed in into console we can retrieve some key information on the Gene Model build as well as summary metrics


```r
TxDb.Hsapiens.UCSC.hg19.knownGene
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
```

---
# Accessing information.

As suggested in the name, **TxDb** objects are a database containing information we would store in GTF/GFF files.

To access information in our **TxDb** object we will need to use special accessor functions.

These accessors return information in **GRanges'** formats allowing us to work with methods from the **GenomicRanges** packages
---
# Extracting Genomic locations.

We have a number of functions which allow us to extract GRanges for our features of interest including

* **genes()** - Extract gene locations as a **GRanges**.
* **transcripts()** - Extract transcript locations  as a **GRanges**.
* **exons()** - Extract exon locations  as a **GRanges**.
* **cds()** - Extract coding locations  as a **GRanges**.
* **promoters()** - Extract promoter locations  as a **GRanges**.


---
# Extracting genes

We can extract genes with the **genes()** function and the **TxDb** object as a parameter.


```r
myGenes &lt;- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
myGenes
```

```
## GRanges object with 23056 ranges and 1 metadata column:
##         seqnames                 ranges strand |     gene_id
##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19 [ 58858172,  58874214]      - |           1
##      10     chr8 [ 18248755,  18258723]      + |          10
##     100    chr20 [ 43248163,  43280376]      - |         100
##    1000    chr18 [ 25530930,  25757445]      - |        1000
##   10000     chr1 [243651535, 244006886]      - |       10000
##     ...      ...                    ...    ... .         ...
##    9991     chr9 [114979995, 115095944]      - |        9991
##    9992    chr21 [ 35736323,  35743440]      + |        9992
##    9993    chr22 [ 19023795,  19109967]      - |        9993
##    9994     chr6 [ 90539619,  90584155]      + |        9994
##    9997    chr22 [ 50961997,  50964905]      - |        9997
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting exons

We can extract exons with the **exons()** function and the **TxDb** object as a parameter in the same way.


```r
myExons &lt;- exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
myExons
```

```
## GRanges object with 289969 ranges and 1 metadata column:
##                  seqnames         ranges strand |   exon_id
##                     &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 [11874, 12227]      + |         1
##        [2]           chr1 [12595, 12721]      + |         2
##        [3]           chr1 [12613, 12721]      + |         3
##        [4]           chr1 [12646, 12697]      + |         4
##        [5]           chr1 [13221, 14409]      + |         5
##        ...            ...            ...    ... .       ...
##   [289965] chrUn_gl000241 [35706, 35859]      - |    289965
##   [289966] chrUn_gl000241 [36711, 36875]      - |    289966
##   [289967] chrUn_gl000243 [11501, 11530]      + |    289967
##   [289968] chrUn_gl000243 [13608, 13637]      + |    289968
##   [289969] chrUn_gl000247 [ 5787,  5816]      - |    289969
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting transcripts

We can extract transcripts with the **transcripts()** functions and the **TxDb** object as a parameter.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
myTranscripts
```

```
## GRanges object with 82960 ranges and 2 metadata columns:
##                 seqnames           ranges strand |     tx_id     tx_name
##                    &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##       [1]           chr1 [ 11874,  14409]      + |         1  uc001aaa.3
##       [2]           chr1 [ 11874,  14409]      + |         2  uc010nxq.1
##       [3]           chr1 [ 11874,  14409]      + |         3  uc010nxr.1
##       [4]           chr1 [ 69091,  70008]      + |         4  uc001aal.1
##       [5]           chr1 [321084, 321115]      + |         5  uc001aaq.2
##       ...            ...              ...    ... .       ...         ...
##   [82956] chrUn_gl000237   [    1,  2686]      - |     82956  uc011mgu.1
##   [82957] chrUn_gl000241   [20433, 36875]      - |     82957  uc011mgv.2
##   [82958] chrUn_gl000243   [11501, 11530]      + |     82958  uc011mgw.1
##   [82959] chrUn_gl000243   [13608, 13637]      + |     82959  uc022brq.1
##   [82960] chrUn_gl000247   [ 5787,  5816]      - |     82960  uc022brr.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting Coding regions

We can extract coding regions with the **cds()** functions and the **TxDb** object as a parameter.


```r
myCDS &lt;- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)
myCDS
```

```
## GRanges object with 237533 ranges and 1 metadata column:
##                  seqnames           ranges strand |    cds_id
##                     &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 [ 12190,  12227]      + |         1
##        [2]           chr1 [ 12595,  12721]      + |         2
##        [3]           chr1 [ 13403,  13639]      + |         3
##        [4]           chr1 [ 69091,  70008]      + |         4
##        [5]           chr1 [324343, 324345]      + |         5
##        ...            ...              ...    ... .       ...
##   [237529] chrUn_gl000228   [30530, 31035]      - |    237529
##   [237530] chrUn_gl000228   [31353, 31430]      - |    237530
##   [237531] chrUn_gl000228   [31660, 31734]      - |    237531
##   [237532] chrUn_gl000228   [31660, 31737]      - |    237532
##   [237533] chrUn_gl000228   [31996, 32173]      - |    237533
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting additional information for exons/transcripts/cds regions.

We can control which additional information such as transcript ID and gene ID we retrieve in metadata columns using the **columns** parameter in the **exons()/transcripts()/cds()** functions. 

For a full list of available columsn see **?transcripts()** function help.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             columns=c("gene_id","tx_id"))
myTranscripts[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames         ranges strand |         gene_id     tx_id
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;integer&gt;
##   [1]     chr1 [11874, 14409]      + |       100287102         1
##   [2]     chr1 [11874, 14409]      + |       100287102         2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting Promoter regions

Another useful function is the **promoters()** function.

This accepts additional parameters of **upstream** and **downstream** to specify regions around TSS to be used as a promoter while accounting for a transcripts's strand.

We have been using the **resize()** function with **GRanges** objects for the same purpose.


```r
myPromoters &lt;- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             upstream=2000,downstream=50)
myPromoters[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames        ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr1 [9874, 11923]      + |         1  uc001aaa.3
##   [2]     chr1 [9874, 11923]      + |         2  uc010nxq.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

A major component of gene models in GFF and GTF files is the grouping of exons to transcripts/genes and transcripts to genes.

We can extract lists of GRanges (a GRangesList) containing the grouping of features to metafeatures using additional accessor functions.

These include the **transcriptsBy()**, **exonsBy()** and **cdsBy()**


---
# Extracting groups of genomic features.

The **transcriptsBy()** function takes as argument a **TxDb** object and a **by** argument specifying the metafeature/feature to group by.

For the **transcriptsBy()**, the **by** argument may be gene, exon or cds


```r
transcriptByGenes &lt;- transcriptsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="gene")
transcriptByGenes[1:2]
```

```
## GRangesList object of length 2:
## $1 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames               ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 [58858172, 58864865]      - |     70455  uc002qsd.4
##   [2]    chr19 [58859832, 58874214]      - |     70456  uc002qsf.2
## 
## $10 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames               ranges strand | tx_id    tx_name
##   [1]     chr8 [18248755, 18258723]      + | 31944 uc003wyw.1
## 
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

As with all lists we can extract information from elements in the GRangesList using the **$** and **[[]]** accessors.

This will extract the grouped intervals as a GRanges.


```r
# transcriptByGenes$1 or
transcriptByGenes[[1]]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames               ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 [58858172, 58864865]      - |     70455  uc002qsd.4
##   [2]    chr19 [58859832, 58874214]      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

Similarly the very useful **exonsBy()** function takes again an argument for a **TxDb** object and a **by** argument.

For the **exonsBy()** and **cdsBy()** functions, the **by** argument may be gene, or tx


```r
exonsByTranscript &lt;- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="tx")
exonsByTranscript[1:2]
```

```
## GRangesList object of length 2:
## $1 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]     chr1 [11874, 12227]      + |         1        &lt;NA&gt;         1
##   [2]     chr1 [12613, 12721]      + |         3        &lt;NA&gt;         2
##   [3]     chr1 [13221, 14409]      + |         5        &lt;NA&gt;         3
## 
## $2 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand | exon_id exon_name exon_rank
##   [1]     chr1 [11874, 12227]      + |       1      &lt;NA&gt;         1
##   [2]     chr1 [12595, 12721]      + |       2      &lt;NA&gt;         2
##   [3]     chr1 [13403, 14409]      + |       6      &lt;NA&gt;         3
## 
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome
```


---
# Creating TxDb from external sources.

Often we will have our favourite gene models stored in a GTF outside of Bioconductor objects. If we wish to take advantage of all the features in Bioconductor we will need to import it into a **TxDb** object.

We can do this using the **makeTxDbFromGFF()** function which will parse GTF or GFF into a **TxDb** object

```r
myCustomTxDb &lt;- makeTxDbFromGFF("../../Data/Xkr4.gtf")
```

```
## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
```

```r
class(myCustomTxDb)
```

```
## [1] "TxDb"
## attr(,"package")
## [1] "GenomicFeatures"
```

---

```r
myCustomTxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: ../../Data/Xkr4.gtf
## # Organism: NA
## # Taxonomy ID: NA
## # miRBase build ID: NA
## # Genome: NA
## # transcript_nrow: 1
## # exon_nrow: 3
## # cds_nrow: 3
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2018-04-13 09:34:23 -0400 (Fri, 13 Apr 2018)
## # GenomicFeatures version at creation time: 1.28.4
## # RSQLite version at creation time: 2.0
## # DBSCHEMAVERSION: 1.1
```

---
# Creating TxDb from external sources.

Now we have our own custom GTF object we can make use of the functions from **GenomicFeatures** package such as retrieving gene coordinates as a **GRanges**.


```r
genes(myCustomTxDb)
```

```
## GRanges object with 1 range and 1 metadata column:
##        seqnames             ranges strand |     gene_id
##           &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   Xkr4     chr1 [3214482, 3671498]      - |        Xkr4
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
# Creating TxDb from external sources.

Or exons grouped by genes as a **GRangesList**



```r
exonsBy(myCustomTxDb,by="gene")
```

```
## GRangesList object of length 1:
## $Xkr4 
## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames             ranges strand |   exon_id   exon_name
##          &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr1 [3214482, 3216968]      - |         1        &lt;NA&gt;
##   [2]     chr1 [3421702, 3421901]      - |         2        &lt;NA&gt;
##   [3]     chr1 [3670552, 3671498]      - |         3        &lt;NA&gt;
## 
## -------
## seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
# Creating TxDb from UCSC.

We can create a TxDb directly from UCSC's database using the **makeTxDbFromUCSC()** function.

The **makeTxDbFromUCSC** function simply requires a genome name to build a **TxDb** object. 

We can review the genomes available to us using the **ucscGenomes()** function from the **rtracklayer** package. This returns a data frame of genomes from which we can build a TxDB object.



```r
library(rtracklayer)
availableGenomes &lt;- ucscGenomes()
availableGenomes[1:4,]
```

```
##     db species      date                               name
## 1 hg38   Human Dec. 2013 Genome Reference Consortium GRCh38
## 2 hg19   Human Feb. 2009 Genome Reference Consortium GRCh37
## 3 hg18   Human Mar. 2006                    NCBI Build 36.1
## 4 hg17   Human  May 2004                      NCBI Build 35
```
---
# Creating TxDb from UCSC.

Once we have identified the genome build we wish to use we can specify the genome parameter in the **makeTxDbFromUCSC()** function.


```r
hg18TxDb &lt;- makeTxDbFromUCSC(genome="hg18")
hg18TxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg18
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # UCSC Track: UCSC Genes
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: NA
## # transcript_nrow: 66803
## # exon_nrow: 266688
## # cds_nrow: 221991
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2018-04-13 09:36:10 -0400 (Fri, 13 Apr 2018)
## # GenomicFeatures version at creation time: 1.28.4
## # RSQLite version at creation time: 2.0
## # DBSCHEMAVERSION: 1.1
```
---
# Creating TxDb from UCSC.

And again we can now make use of **GenomicFeatures** package's functions with our new TxDb objects


```r
hg18Promoters &lt;- promoters(hg18TxDb,2000,50)
```
---
# Exporting a TxDb object as a GTF/GFF.

We can make use of the **export.gff()** in the **rtracklayer** package to export our **TxDb** object to a gtf or gff file. We can include an additional argument of **format** to the **export.gff()** function.

We can specify the **format** to be **gff** or **gtf** as shown below.


```r
export.gff(myCustomTxDb,con="customTxDbb.gff",format="gff")
export.gff(myCustomTxDb,con="customTxDbb.gtf",format="gtf")
```
---
# Some useful functions for working with gene model in GenomicFeatures.

We have seen that we can extract transcripts and exons grouped by their genes/transcripts using the **transcriptsBy()** and **exonsBy()** functions.

To identify the total number of groups we can simply use the **length()** function. Here we have 23459 transcipts or groups of exons.


```r
transcriptByGenes &lt;- exonsBy(hg18TxDb,by="gene")
length(transcriptByGenes)
```

```
## [1] 20121
```
---
# Some useful functions for working with gene model in GenomicFeatures.

To find the number of exons in every group/transcript we can use the function **lengths()**. The **lengths()** function will tell us the length of every group in our **GRangesList**


```r
transcriptNumberPerGene &lt;- lengths(transcriptByGenes)
transcriptNumberPerGene[1:5]
```

```
##     1    10   100  1000 10000 
##     2     1     1     2     3
```
---
# Some useful functions for working with gene model in GenomicFeatures.

We may wish to know the total sum length of all exons in every transcript. The **transcriptLengths()** function allows us to find all transcript lengths from a TxDb object as well as additional information.


```r
transcript_Lens &lt;- transcriptLengths(hg18TxDb)
transcript_Lens[1:5,]
```

```
##   tx_id    tx_name gene_id nexon tx_len
## 1     1 uc001aaa.2    &lt;NA&gt;     3   2122
## 2     2 uc009vip.1    &lt;NA&gt;     2   2772
## 3     3 uc009vjg.1    &lt;NA&gt;     3    708
## 4     4 uc009vjh.1   79501     3    888
## 5     5 uc001aal.1   79501     1    918
```
---
# Extracting sequence information.

As with many Bioconductor packages, there is interoperability between **BSgenome** packages holding genome sequences and **TxDb** objects with gene models. One very useful function is the **extractTranscriptSeqs()** function.

The **extractTranscriptSeqs()** accepts a BSgenome object of genomic sequence of interest and a TxDb object.


```r
library(BSgenome.Hsapiens.UCSC.hg19)
hg19TransSeq &lt;- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, 
                                      transcripts=TxDb.Hsapiens.UCSC.hg19.knownGene)
hg19TransSeq
```

```
##   A DNAStringSet instance of length 82960
##          width seq                                     names               
##     [1]   1652 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 1
##     [2]   1488 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 2
##     [3]   1595 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 3
##     [4]    918 ATGGTGACTGAATTCATT...TCTAGTGTAAAGTTTTAG 4
##     [5]     32 TACAGACCAAGCTCATGACTCACAATGGCCTA        5
##     ...    ... ...
## [82956]   1217 GCCAGTTTAGGGTCTCTG...CCTCCGCCTCCTGAGATC 82956
## [82957]    737 CTCCACTTCTGATCCTCC...GCTTTATTAGATGCAGTG 82957
## [82958]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA          82958
## [82959]     30 TTGCAGAGGTGGCTGGTTGCTCTTTGAGCC          82959
## [82960]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA          82960
```
---
# Extracting sequence information.

We can then either work with the resulting DNAStringSet then in R or write to Fasta file for analysis in external programs.


```r
writeXStringSet(hg19TransSeq,"myTranscriptSequences.fa")
```
---
# Gene annotation.

Where as the **TxDb** packages hold information on gene models such as exons positions and exon to gene mapping, the **OrgDb** packages contain information on gene's mapping to alternative IDs and any functional annotation. 

Information on model organism's gene annotation is contained with the **org.db** packages.

Format is org. **species** . **ID type** .db

Homo Sapiens annotation with Entrez Gene IDs -- org.Hs.eg.db

&lt;div align="center"&gt;
&lt;img src="imgs/orgdb.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;

---
# Gene annotation.

To gain access to gene annotation we must first load the library for the relevant species.

Here we install and load the library for Human, Org.Hs.eg.db.



```r
biocLite("org.Hs.eg.db")
library(org.Hs.eg.db)
class(org.Hs.eg.db)
```


```
## [1] "OrgDb"
## attr(,"package")
## [1] "AnnotationDbi"
```

---
## Org.db AnnotationDbi Accessor Functions

As with the TxDb packages, we will need to use special accessors to retrieve information from Org.db.

We can use the functions:-

- **columns()** - Display what kind of annotation is available in OrgDb objects.
- **keytypes()** - Displays which type of identifiers can be used with **select** function. 
- **keys()** - Returns keys (index) for the database contained in the OrgDb object. Used along with **keytypes()** in &lt;b&gt;select&lt;/b&gt; function to retrieve interested annotation
- **select** - Retrieve the annotation data as a data.frame based on the supplied keys, keytypes and columns. 

We will explore how to retrieve annotation from gene-centric organism level annotation package (org.Hs.eg.db) using these functions.

---
## Accessing annotation from OrgDb

We can use the **columns()** function first to list available annotation.

Here we see our annotation for alternative IDs (ENSEMBL, REFSEQ, GENENAME) as well as functional annotation (GO, PATH, OMIM)


```r
columns(org.Hs.eg.db)
```

```
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" 
##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    
##  [9] "EVIDENCEALL"  "GENENAME"     "GO"           "GOALL"       
## [13] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    
## [17] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [25] "UNIGENE"      "UNIPROT"
```
--
To know more about the above identifier types

```r
help(GENENAME)
```
---
## Accessing annotation from org.Hs.eg.db {.smaller}

Which keytypes can be used to query this database can be retrieved using keytypes.

```r
keytypes(org.Hs.eg.db)
```

```
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT" 
##  [5] "ENSEMBLTRANS" "ENTREZID"     "ENZYME"       "EVIDENCE"    
##  [9] "EVIDENCEALL"  "GENENAME"     "GO"           "GOALL"       
## [13] "IPI"          "MAP"          "OMIM"         "ONTOLOGY"    
## [17] "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
## [25] "UNIGENE"      "UNIPROT"
```
---

## Accessing annotation from OrgDb

If we want to extract few identifiers of a particular keytype, we can use keys() function

```r
keys(org.Hs.eg.db, keytype="SYMBOL")
```

```
##    [1] "A1BG"       "A2M"        "A2MP1"      "NAT1"       "NAT2"      
##    [6] "NATP"       "SERPINA3"   "AADAC"      "AAMP"       "AANAT"     
##   [11] "AARS"       "AAVS1"      "ABAT"       "ABCA1"      "ABCA2"     
##   [16] "ABCA3"      "ABCB7"      "ABCF1"      "ABCA4"      "ABL1"      
##   [21] "AOC1"       "ABL2"       "ABO"        "ABR"        "ACAA1"     
##   [26] "ACACA"      "ACACB"      "ACADL"      "ACADM"      "ACADS"     
##   [31] "ACADSB"     "ACADVL"     "ACAT1"      "ACAT2"      "ASIC2"     
##   [36] "ASIC1"      "ACHE"       "ACLS"       "ACLY"       "ACO1"      
##   [41] "ACR"        "ACO2"       "ACOX1"      "ACP1"       "ACP2"      
##   [46] "ACP5"       "ACPP"       "ACRV1"      "ACTA1"      "ACTA2"     
##   [51] "ACTB"       "ACTBP2"     "ACTBP3"     "ACTBP4"     "ACTBP6"    
##   [56] "ACTBP7"     "ACTBP8"     "ACTBP9"     "ACTC1"      "ACTG1"     
##   [61] "ACTG2"      "ACTG1P1"    "ACTG1P2"    "ACTG1P3"    "ACTG1P6"   
##   [66] "ACTG1P7"    "ACTG1P8"    "ACTN4"      "ACTG1P9"    "ACTG1P10"  
##   [71] "ACTL6A"     "ACTN1"      "ACTN2"      "ACTN3"      "ACVR1"     
##   [76] "ACVR1B"     "ACVR2A"     "ACVR2B"     "ACVRL1"     "ACY1"      
##   [81] "ACYP1"      "ACYP2"      "ADA"        "ADAM8"      "ADAM10"    
##   [86] "ADAR"       "ADARB1"     "ADARB2"     "ADCP1"      "ADCY1"     
##   [91] "ADCY2"      "ADCY3"      "ADCY5"      "ADCY6"      "ADCY7"     
##   [96] "ADCY8"      "ADCY9"      "ADCYAP1"    "ADCYAP1R1"  "ADD1"      
##  [101] "ADD2"       "ADD3"       "ADFN"       "PLIN2"      "ADH1A"     
##  [106] "ADH1B"      "ADH1C"      "ADH4"       "ADH5"       "ADH6"      
##  [111] "ADH7"       "ADK"        "ADM"        "ADORA1"     "ADORA2A"   
##  [116] "ADORA2B"    "ADORA2BP1"  "ADORA3"     "ADPRH"      "PARP1"     
##  [121] "PARP4"      "PARP1P1"    "PARP1P2"    "ADRA1D"     "ADRA1B"    
##  [126] "ADRA1A"     "ADRA2A"     "ADRA2B"     "ADRA2C"     "ADRB1"     
##  [131] "ADRB2"      "ADRB3"      "GRK2"       "GRK3"       "ADSL"      
##  [136] "ADSS"       "AP2A1"      "AP2A2"      "AP1B1"      "AP2B1"     
##  [141] "AP1G1"      "AEBP1"      "AES"        "CRISP1"     "AF8T"      
##  [146] "AFA"        "AFD1"       "AFG3L1P"    "AFM"        "AFP"       
##  [151] "AGA"        "ACAN"       "AGER"       "AGL"        "AGMX2"     
##  [156] "AGRP"       "JAG1"       "AGT"        "AGTR1"      "AGTR2"     
##  [161] "APLNR"      "AGXT"       "NR0B1"      "AHCY"       "AIC"       
##  [166] "AHR"        "AHSG"       "AIF1"       "AIH3"       "CRYBG1"    
##  [171] "AK1"        "AK2"        "AK4"        "AK4P1"      "AKT1"      
##  [176] "AKT2"       "ALAD"       "ALAS1"      "ALAS2"      "ALB"       
##  [181] "ALCAM"      "ABCD1"      "ALDH1A1"    "ALDH2"      "ALDH3A1"   
##  [186] "ALDH1B1"    "ALDH1A3"    "ALDH3B1"    "ALDH3B2"    "ALDH9A1"   
##  [191] "ALDH3A2"    "ABCD2"      "ALDOA"      "ALDOAP2"    "ALDOB"     
##  [196] "ALDOC"      "AKR1B1"     "AKR1B1P2"   "ALK"        "ALOX12"    
##  [201] "ALOX5"      "ALOX5AP"    "ALOX12B"    "ALOX12P1"   "ALOX12P2"  
##  [206] "ALOX15"     "ALOX15B"    "ALPI"       "ALPL"       "ALPP"      
##  [211] "ALPPL2"     "ALS3"       "ALS5"       "ALX3"       "AMBN"      
##  [216] "AMBP"       "AMCN"       "AMD1"       "AMD1P2"     "AMELX"     
##  [221] "AMELY"      "AMFR"       "AMH"        "AMHR2"      "AMPD1"     
##  [226] "AMPD2"      "AMPD3"      "AMPH"       "BIN1"       "AMT"       
##  [231] "AMY1A"      "AMY1B"      "AMY1C"      "AMY2A"      "AMY2B"     
##  [236] "AMYP1"      "ANCR"       "ANG"        "ANGPT1"     "ANGPT2"    
##  [241] "ANK1"       "ANK2"       "ANK3"       "ANOP1"      "ANPEP"     
##  [246] "SLC25A4"    "SLC25A5"    "SLC25A6"    "ANXA1"      "ANXA2"     
##  [251] "ANXA2P1"    "ANXA2P2"    "ANXA2P3"    "ANXA3"      "ANXA4"     
##  [256] "ANXA5"      "ANXA6"      "ANXA7"      "ANXA11"     "ANXA13"    
##  [261] "AOAH"       "AOC2"       "AOX1"       "APAF1"      "NUDT2"     
##  [266] "APOF"       "APBA1"      "APBA2"      "APBB1"      "APBB2"     
##  [271] "APC"        "APCS"       "AIRE"       "APEH"       "APEX1"     
##  [276] "BIRC2"      "BIRC3"      "XIAP"       "BIRC5"      "APLP1"     
##  [281] "APLP2"      "APOA1"      "APOA2"      "APOA4"      "APOB"      
##  [286] "APOBEC1"    "APOC1"      "APOC1P1"    "AQP8"       "APOC2"     
##  [291] "APOC3"      "APOC4"      "APOD"       "APOE"       "APOH"      
##  [296] "APP"        "APRT"       "KLK3"       "FAS"        "FASLG"     
##  [301] "SHROOM2"    "AQP1"       "AQP2"       "AQP3"       "AQP4"      
##  [306] "AQP5"       "AQP6"       "AQP7"       "AQP9"       "AR"        
##  [311] "ABCC6"      "ARAF"       "ARCN1"      "TRIM23"     "AREG"      
##  [316] "ARF1"       "ARF3"       "ARF4"       "ARL4D"      "ARF4P1"    
##  [321] "ARF5"       "ARF6"       "ARG1"       "ARG2"       "RHOA"      
##  [326] "RHOB"       "RHOC"       "RND3"       "RHOG"       "ARHGAP1"   
##  [331] "ARHGAP4"    "ARHGAP5"    "ARHGAP6"    "ARHGDIA"    "ARHGDIB"   
##  [336] "ARHGDIG"    "RHOH"       "ARL1"       "PHOX2A"     "ARL2"      
##  [341] "ARL3"       "ARNT"       "ARNTL"      "ARR3"       "ARRB1"     
##  [346] "ARRB2"      "ARSA"       "ARSB"       "STS"        "ARSC2"     
##  [351] "ARSD"       "ARSE"       "ARSF"       "ART1"       "ART2P"     
##  [356] "ART3"       "ART4"       "ARVCF"      "ARVD3"      "ARVD4"     
##  [361] "ASAH1"      "ASCL1"      "ASCL2"      "ASD1"       "ASGR1"     
##  [366] "ASGR2"      "ASIP"       "ASL"        "ASLP1"      "ASMT"      
##  [371] "ASNA1"      "ASNS"       "ASNSP2"     "ASPA"       "ASPH"      
##  [376] "ASS1"       "ASS1P2"     "ASS1P4"     "ASS1P5"     "ASS1P6"    
##  [381] "ASS1P8"     "ASS1P9"     "ASS1P10"    "ASS1P14"    "ASTN1"     
##  [386] "SERPINC1"   "ZFHX3"      "ATD"        "ATF1"       "ATF3"      
##  [391] "ATF4"       "ATF4P1"     "ATHS"       "ATIC"       "ATM"       
##  [396] "RERE"       "ATOH1"      "ATOX1"      "ATP1A1"     "ATP1A2"    
##  [401] "ATP1A3"     "ATP12A"     "ATP1A4"     "ATP1B1"     "ATP1B2"    
##  [406] "ATP1B3"     "ATP1B3P1"   "ATP1B1P1"   "FXYD2"      "ATP2A1"    
##  [411] "ATP2A2"     "ATP2A3"     "ATP2B1"     "ATP2B2"     "ATP2B3"    
##  [416] "ATP2B4"     "ATP4A"      "ATP4B"      "ATP5A1"     "ALDH7A1"   
##  [421] "ATP5A1P2"   "ATP5B"      "ATP5BP1"    "ATP5C1"     "ATP5D"     
##  [426] "ATP5E"      "ATP5F1"     "ATP5G1"     "ATP5G2"     "ATP5G3"    
##  [431] "ATP5G2P1"   "ATP5I"      "ATP5J"      "ATP6V1A"    "ATP6V1B1"  
##  [436] "ATP6V1B2"   "ATP6V0C"    "ATP6V1C1"   "ATP6V1E1"   "ATP6V0B"   
##  [441] "ATP6V1G2"   "ATP6V0A1"   "ATP6AP1"    "ATP7A"      "ATP5O"     
##  [446] "ATP7B"      "ALDH7A1P1"  "ALDH7A1P2"  "ALDH7A1P3"  "ALDH7A1P4" 
##  [451] "ATR"        "ATRX"       "KIF1A"      "AUH"        "AUP1"      
##  [456] "AVP"        "AVPR1A"     "AVPR1B"     "AVPR2"      "AXL"       
##  [461] "AZF1"       "AZGP1"      "AZU1"       "B2M"        "B2MR"      
##  [466] "BAAT"       "BACH1"      "BAD"        "BAG1"       "BAGE"      
##  [471] "ADGRB1"     "ADGRB2"     "ADGRB3"     "BAK1"       "NKX3-2"    
##  [476] "BARD1"      "BAX"        "BBS1"       "BBS2"       "BBS4"      
##  [481] "BCAT1"      "BCAT2"      "BCHE"       "BCKDHA"     "BCKDHB"    
##  [486] "CCND1"      "BCL2"       "BCL2A1"     "BCL2L1"     "BCL2L2"    
##  [491] "BAK1P1"     "BAK1P2"     "BCL3"       "BCL5"       "BCL6"      
##  [496] "BCL7A"      "NBEAP1"     "BCL9"       "TNFRSF17"   "HCN2"      
##  [501] "OPN1SW"     "BCR"        "BCRP4"      "BCS1L"      "BCYRN1"    
##  [506] "BCYRN1P1"   "BCYRN1P2"   "BDH1"       "BDKRB1"     "BDKRB2"    
##  [511] "BDNF"       "BEVI"       "CFB"        "BFSP1"      "BGLAP"     
##  [516] "BGN"        "CEACAM1"    "BHMT"       "BICD1"      "BID"       
##  [521] "BIK"        "PRDM1"      "BLK"        "BLM"        "BLMH"      
##  [526] "CXCR5"      "BLVRA"      "BLVRB"      "BNC1"       "BMI1"      
##  [531] "BMP1"       "BMP2"       "BMP3"       "BMP4"       "BMP5"      
##  [536] "BMP6"       "BMP7"       "BMP8B"      "BMPR1A"     "BMPR1B"    
##  [541] "BMPR2"      "BMX"        "POLR3D"     "BNIP1"      "BNIP2"     
##  [546] "BNIP3"      "BNIP3L"     "BOK"        "DST"        "FOXL2"     
##  [551] "BPGM"       "BPHL"       "BPI"        "BRCA1"      "BRAF"      
##  [556] "BRCA2"      "BRDT"       "ZFP36L1"    "ZFP36L2"    "BRS3"      
##  [561] "BSG"        "BST1"       "BST2"       "BTC"        "BTD"       
##  [566] "KLF9"       "KLF5"       "BTF3"       "BTF3P11"    "BTF3P1"    
##  [571] "BTG1"       "BTK"        "BTN1A1"     "BTS1"       "BUB1"      
##  [576] "BVR1"       "BUB1B"      "BYSL"       "TSPO"       "C1QBP"     
##  [581] "SERPING1"   "ERC2-IT1"   "C1QA"       "C1QB"       "C1QC"      
##  [586] "C1R"        "C1S"        "C2"         "C3"         "C3AR1"     
##  [591] "C4A"        "C4B"        "C4BPA"      "C4BPAP2"    "C4BPB"     
##  [596] "CAPN5"      "C5"         "C5AR1"      "C6"         "C7"        
##  [601] "C8A"        "C8B"        "C8G"        "OSGIN2"     "C9"        
##  [606] "VPS51"      "MRPL49"     "ZNHIT2"     "MPPED2"     "MYRF"      
##  [611] "TMEM258"    "DAGLA"      "GAS8-AS1"   "FMNL1"      "LDLRAD4"   
##  [616] "PTTG1IP"    "C21orf2"    "TMEM50B"    "MPPED1"     "CA1"       
##  [621] "CA2"        "CA3"        "CA4"        "CA5A"       "CA5AP1"    
##  [626] "CA6"        "CA7"        "CA8"        "CA9"        "CA11"      
##  [631] "CA12"       "CACD"       "CACNA1A"    "CACNA1B"    "CACNA1C"   
##  [636] "CACNA1D"    "CACNA1E"    "CACNA1F"    "CACNA1S"    "DDR1"      
##  [641] "CACNA2D1"   "CACNB1"     "CACNB2"     "CACNB3"     "CACNB4"    
##  [646] "CACNG1"     "SLC25A20"   "SLC25A20P1" "CAD"        "CALB1"     
##  [651] "CALB2"      "S100G"      "CALCA"      "CALCB"      "CALCP"     
##  [656] "CALCR"      "CALD1"      "CALM1"      "CALM1P1"    "CALM1P2"   
##  [661] "CALM2"      "CALM2P2"    "CALM2P3"    "CALM3"      "CALML3"    
##  [666] "CALR"       "CALU"       "CAMK4"      "CAMK2A"     "CAMK2B"    
##  [671] "CAMK2D"     "CAMK2G"     "CAMLG"      "CAMP"       "CANX"      
##  [676] "CAPG"       "CAPN1"      "CAPN2"      "CAPN3"      "CAPNS1"    
##  [681] "CAPN6"      "CAPS"       "CAPZA1"     "CAPZA2"     "CAST"      
##  [686] "CAPZB"      "CARS"       "CASP1"      "CASP2"      "CASP3"     
##  [691] "CASP4"      "CASP5"      "CASP6"      "CASP7"      "CASP8"     
##  [696] "CASP9"      "CASP10"     "CASQ1"      "CASQ2"      "CASR"      
##  [701] "CAT"        "CATR1"      "CAV1"       "CAV2"       "CAV3"      
##  [706] "RUNX2"      "RUNX1"      "RUNX1T1"    "CBFA2T3"    "RUNX3"     
##  [711] "CBFB"       "SERPINA6"   "CBL"        "CBLB"       "CBLN1"     
##  [716] "SERPINH1"   "CBR1"       "CBR3"       "CBS"        "CCA1"      
##  [721] "CCIN"       "CCAL1"      "KYAT1"      "CCK"        "CCKAR"     
##  [726] "CCKBR"      "KRIT1"      "CCNA2"      "CCNB1"      "CCNC"      
##  [731] "CCND2"      "CCND2P1"    "CCND3"      "CCND3P1"    "CCNE1"     
##  [736] "CCNF"       "CCNG1"      "CCNG2"      "CCNH"       "CCNT1"     
##  [741] "CCNT2"      "CCT"        "CCT6A"      "CD1A"       "CD1B"      
##  [746] "CD1C"       "CD1D"       "CD1E"       "CD2"        "CD3D"      
##  [751] "CD3E"       "CD3G"       "CD247"      "CD4"        "CD5"       
##  [756] "CD5L"       "CD6"        "CD7"        "CD8A"       "CD8B"      
##  [761] "CD8BP"      "CD9"        "CD14"       "CD19"       "MS4A1"     
##  [766] "MS4A3"      "CD22"       "CD24P1"     "CD24P2"     "CD24P3"    
##  [771] "CD24P4"     "CD27"       "CD28"       "CD80"       "CD86"      
##  [776] "TNFRSF8"    "TNFSF8"     "CD33"       "SIGLEC6"    "CD34"      
##  [781] "CD36"       "SCARB1"     "SCARB2"     "CD37"       "CD38"      
##  [786] "ENTPD1"     "ENTPD2"     "ENTPD6"     "ENTPD3"     "ENTPD5"    
##  [791] "CD40"       "CD40LG"     "CD44"       "CD47"       "CD48"      
##  [796] "CD53"       "CD58"       "CD59"       "CD63"       "CD68"      
##  [801] "CD69"       "CD70"       "CD72"       "CD74"       "CD79A"     
##  [806] "CD79B"      "CD81"       "ADGRE5"     "CD151"      "CDA"       
##  [811] "CDAN3"      "CDB2"       "CDK1"       "CDK11B"     "LRBA"      
##  [816] "CDC5L"      "SEPT7"      "CDC6"       "CDC20"      "CDC25A"    
##  [821] "CDC25B"     "CDC25C"     "CDC27"      "CDC34"      "CDC42"     
##  [826] "CDH1"       "CDH2"       "CDH3"       "CDH4"       "CDH5"      
##  [831] "CDH6"       "CDH7"       "CDH8"       "CDH9"       "CDH10"     
##  [836] "CDH11"      "CDH12"      "CDH12P2"    "CDH13"      "CDH15"     
##  [841] "CDH16"      "CDH17"      "CDH18"      "CDK2"       "CDK3"      
##  [846] "CDK4"       "CDK5"       "CDK6"       "CDK7"       "CDL2"      
##  [851] "CDK8"       "CDK9"       "CDKN1A"     "CDKN1B"     "CDKN1C"    
##  [856] "CDKN2A"     "CDKN2B"     "CDKN2C"     "CDKN2D"     "CDKN3"     
##  [861] "CDO1"       "CDR1"       "CDR2"       "CDS1"       "CDSN"      
##  [866] "CD52"       "CDX1"       "CDX2"       "CDX4"       "CLGN"      
##  [871] "CEACAM5"    "CEBPA"      "CEBPB"      "CEBPD"      "CEBPE"     
##  [876] "CEBPG"      "CECR"       "CEL"        "CELP"       "CENPA"     
##  [881] "CENPB"      "CENPC"      "CENPCP1"    "CENPE"      "CENPF"     
##  [886] "CES1"       "CETN1"      "CETN2"      "CETN3"      "CETP"      
##  [891] "CFL1"       "CFL2"       "CFL1P5"     "CTSC"       "CFTR"      
##  [896] "CGA"        "CGB3"       "CGF1"       "CEACAM3"    "CEACAMP5"  
##  [901] "CEACAMP11"  "CEACAM7"    "CEACAM8"    "CEACAM4"    "CEACAMP1"  
##  [906] "CEACAMP2"   "CEACAMP3"   "CEACAMP4"   "CEACAMP6"   "CEACAMP7"  
##  [911] "CEACAMP8"   "CEACAMP9"   "CEACAMP10"  "CHAD"       "RCBTB2"    
##  [916] "CHAT"       "RCC1"       "CHD1"       "CHD2"       "CHD3"      
##  [921] "CHD4"       "AKR1C4"     "CHEK1"      "FOXN3"      "CHGA"      
##  [926] "CHGB"       "CHI3L1"     "CHI3L2"     "CHIT1"      "CHKA"      
##  [931] "CHKB"       "CHM"        "CHML"       "CHN1"       "CHN2"      
##  [936] "CHR"        "CHRM1"      "CHRM2"      "LYST"       "CHRM3"     
##  [941] "CHRM4"      "CHRM5"      "CHRNA1"     "CHRNA2"     "CHRNA3"    
##  [946] "CHRNA4"     "CHRNA5"     "CHRNA7"     "CHRNB1"     "CHRNB2"    
##  [951] "CHRNB3"     "CHRNB4"     "CHRND"      "CHRNE"      "CHRNG"     
##  [956] "CHUK"       "CIDEA"      "CKB"        "CIRBP"      "CISH"      
##  [961] "TBCB"       "CKBE"       "CKBP1"      "CKM"        "CKMT1B"    
##  [966] "CKMT2"      "ERCC8"      "CKS1B"      "CKS2"       "SCAR2"     
##  [971] "CLA3"       "AP2M1"      "AP1S1"      "AP2S1"      "AP3S1"     
##  [976] "CLC"        "CLCA1"      "CLCN1"      "CLCN2"      "CLCN3"     
##  [981] "CLCN4"      "CLCN5"      "CLCN6"      "CLCN7"      "CLCNKA"    
##  [986] "CLCNKB"     "CLU"        "CLIC1"      "CLIC2"      "CLK1"      
##  [991] "CLK2"       "CLK2P1"     "CLK3"       "TPP1"       "CLN3"      
##  [996] "CLN5"       "CLNS1AP1"   "CLNS1A"     "CLPS"       "CLPTM1"    
##  [ reached getOption("max.print") -- omitted 58682 entries ]
```
---
## Accessing annotation from OrgDb

We can extract other annotations for a particular identifier using &lt;b&gt;select&lt;/b&gt; function

```r
select(org.Hs.eg.db, keys = "A1BG", keytype = "SYMBOL", 
       columns = c("SYMBOL", "GENENAME", "CHR") )
```

```
##   SYMBOL               GENENAME CHR
## 1   A1BG alpha-1-B glycoprotein  19
```

---
## Annotating results

We can use this now to translate our Entrez IDs from TxDb object to a gene name

```r
geneLocations &lt;- genes(hg18TxDb)
geneLocations
```

```
## GRanges object with 19742 ranges and 1 metadata column:
##         seqnames                 ranges strand |     gene_id
##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19 [ 63549984,  63565932]      - |           1
##      10     chr8 [ 18293035,  18303003]      + |          10
##     100    chr20 [ 42681577,  42713790]      - |         100
##    1000    chr18 [ 23784933,  24011189]      - |        1000
##   10000     chr1 [241718158, 242073207]      - |       10000
##     ...      ...                    ...    ... .         ...
##    9991     chr9 [114020538, 114135733]      - |        9991
##    9992    chr21 [ 34658193,  34665310]      + |        9992
##    9993    chr22 [ 17403795,  17489967]      - |        9993
##    9994     chr6 [ 90596334,  90640876]      + |        9994
##    9997    chr22 [ 49308863,  49310900]      - |        9997
##   -------
##   seqinfo: 49 sequences (1 circular) from hg18 genome
```

---
## Annotating results


```r
IDs &lt;- geneLocations$gene_id
myTable &lt;- select(org.Hs.eg.db, keys = IDs, keytype = "ENTREZID",
                  columns = c("SYMBOL", "GENENAME", "ENTREZID") )
myTable[1:2,]
```

```
##   ENTREZID SYMBOL               GENENAME
## 1        1   A1BG alpha-1-B glycoprotein
## 2       10   NAT2  N-acetyltransferase 2
```
---

# Time for an exercise.

[Link_to_exercises](../../exercises/GenomicFeatures_exercises.html)

[Link_to_answers](../../answers/GenomicFeatures_answers.html)
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
