---
title: "Sequences In Bioconductor <html><div style='float:left'></div><hr color='#EB811B' size=1px width=796px></html>"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/Bioconductor_Introduction/"
output: 
  xaringan::moon_reader:
    css: ["default", "metropolisCustom.css", "metropolis-fontsCustom.css"]
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


# Genomic Sequences

As we have seen earlier, genomic sequences are often stored in FASTA format.

<div align="center">
<img src="imgs/newFASTA.png" alt="igv" height="400" width="800">
</div>


---

# Genomic Sequences in Bioconductor

Genomic sequences can be handled in Bioconductor using the functions in the **Biostrings** package and full genome sequences for many model organisms are contained within the [**BSgenome** packages](https://bioconductor.org/packages/release/BiocViews.html#___AnnotationData).

<div align="center">
<img src="imgs/bsgenomeExample_2020.png" alt="igv" height="415" width="400">
</div>



---

# BSgenome packages

To make use of a BSgenome package we must first install and load the library.

The resulting BSgenome object (here BSgenome.Mmusculus.UCSC.mm10) contains the full genome sequence for the model organism.

```{r load, echo=TRUE,eval=FALSE}
library(BSgenome.Mmusculus.UCSC.mm10)
class(BSgenome.Mmusculus.UCSC.mm10)
```

```{r load1, echo=FALSE,eval=TRUE}
suppressPackageStartupMessages(library(BSgenome.Mmusculus.UCSC.mm10))
class(BSgenome.Mmusculus.UCSC.mm10)
```
---
# BSgenome.Mmusculus.UCSC.mm10
```{r theObject, echo=TRUE,eval=TRUE,collapse=FALSE}
BSgenome.Mmusculus.UCSC.mm10
```
---
# Data in BSgenome objects

Simply by typing the object name we have retrieved some important information.

Importantly it tells us how we can access information in the object.

- **seqnames()** function to retrieve all contig names.
- **$** and **[[]]** accessors to retrieve a given contig's sequence.
---
# Contig Names

We can retrieve contig names using **seqnames()** function.

```{r contignames, echo=TRUE,eval=TRUE,collapse=FALSE}
contigNames <- seqnames(BSgenome.Mmusculus.UCSC.mm10)
contigNames[1:22]
```
---

# Contig Lengths

We can also retrieve contig lengths using **seqlengths()** function.

```{r contiglengths, echo=TRUE,eval=TRUE,collapse=FALSE}
contigLengths <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
contigLengths[1:4]
```

---

# Retrieving Sequence Information

Now we know sequence names, we can retrieve sequence information using either the **$** or **[[]]** accessors.

Here we use the data.frame and list accessor **$** with name of contig of interest
```{r access, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq <- BSgenome.Mmusculus.UCSC.mm10$chr19
chr19_Seq
```
---
# Retrieving Sequence Information


now we use list accessor **[[]]** with name of contig of interest
```{r access2, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq <- BSgenome.Mmusculus.UCSC.mm10[["chr19"]]
chr19_Seq
```

---

# Biostrings and DNAString

The returned sequence object is a **DNAString** object from the **Biostrings** package. DNAStrings are efficient objects for storing and accessing large sequences.

```{r DNAString, echo=TRUE,eval=TRUE,collapse=FALSE}
class(chr19_Seq)
```

**?DNAString** for full help on DNAString objects.
---

# Subsetting a DNAString Set 

We can subset a DNAString object just like indexing a normal vector

```{r DNAStringSub, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq[1:10000000]
```

---

# Subsetting a DNAString Set 

Or we can use the Biostrings function **subseq()** on our DNAString.

Note that **subseq()** is a faster implementation of subsetting for DNAString objects.

```{r DNAStringSub2, echo=TRUE,eval=TRUE,collapse=FALSE}
subseq(chr19_Seq,start=1,end=10000000)
```

---

# Useful Functions

The Biostrings package has many useful functions for handling sequence data.

We can review the frequency of bases using the **alphabetFrequency()** function.

```{r DNAStringSub3, echo=TRUE,eval=TRUE,collapse=FALSE}
alphabetFrequency(chr19_Seq)
```

---

# Useful Functions

We can get the complement,reverse and reverse complement of sequence using the **complement()**, **reverse()**, **reverseComplement()**.

```{r DNAStringSub4, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_SeqComp <- complement(chr19_Seq)
chr19_SeqRev <- reverse(chr19_Seq)
chr19_SeqRevComp <- reverseComplement(chr19_Seq[10000000:10000010])
chr19_Seq[10000000:10000010]
chr19_SeqRevComp
```

---

# Useful Functions

We can get the translation of our sequence using the **translate()**.

```{r DNAStringSub4a, echo=TRUE,eval=TRUE,collapse=FALSE}
length(chr19_Seq[10000000:10000008])
chr19_SeqTranslation <- translate(chr19_Seq[10000000:10000008])
chr19_SeqTranslation
```

---

# Useful Functions

We can get also match and count the occurrence of patterns in our sequence using **matchPattern()** and **countPattern()**.

To count occurrence of patterns we provide our sequence to match to the **pattern** argument and the sequence to search within as a DNAstring object to the **subject** argument.

```{r DNAStringSub5k, echo=TRUE,eval=TRUE,collapse=FALSE,tidy=FALSE}
chr19_Count <- countPattern(pattern="ATCTGCAATG",
                            subject=chr19_Seq)
chr19_Count
```

---

# Useful Functions

We may not always expect a perfect match for our sequence. We can search for close matches by setting the degree of specificity to the **max.mismatch** and **min.mismatch** arguments. 

By default both **max.mismatch** and **min.mismatch**  are set to 0.

```{r DNAStringSubdd5k, echo=TRUE,eval=TRUE,collapse=FALSE,tidy=FALSE}
chr19_Count <- countPattern(pattern="ATCTGCAATG",
                            subject=chr19_Seq,
                            max.mismatch = 2,
                            min.mismatch = 0)
chr19_Count
```

---

# Useful Functions

We can also use IUPAC codes to account for ambiguity in the sequence we will match.

We can see the availble IUPAC code by viewing the Biostring named vector **IUPAC_CODE_MAP**.

```{r DNAStringSub5ssk, echo=TRUE,eval=TRUE,collapse=FALSE,tidy=FALSE}
IUPAC_CODE_MAP
```

---

# Useful Functions

We can make use of IUPAC codes in our search by setting the **fixed** argument to false.

```{r DNAStringSub5ssssk, echo=TRUE,eval=TRUE,collapse=FALSE,tidy=FALSE}
chr19_Count <- countPattern(pattern="RYKHBNKYSRR",
                            subject=chr19_Seq,
                            fixed=FALSE)
chr19_Count
```

---

# Useful Functions

Typically we will search both strands (Watson and Crick/forward and reverse) to identify any patterns. 

```{r DNAStringSub5l, echo=TRUE,eval=TRUE,collapse=FALSE,tidy=FALSE}
chr19_Count_Watson <- countPattern(pattern="ATCTGCAATG",
                                    subject=chr19_Seq)
chr19_Count_Crick <- countPattern(pattern="ATCTGCAATG",
                                    subject=reverseComplement(chr19_Seq)
                                   )
Total_chr19_Count <- chr19_Count_Watson+chr19_Count_Crick

```


---

#Writing to a FASTA File

The **Biostrings** package contains useful functions to read and write to a FASTA file.

To write our DNAString object to a FASTA file we can use the **writeXStringSet()** function. 

First we will convert our DNAString to a DNAStringSet object using the **DNAStringSet()** function and set names using the **names()** function.


```{r DNAStringSub611, echo=TRUE,eval=FALSE,collapse=FALSE}
chr19_SeqSet <- DNAStringSet(chr19_Seq[10000000:10000010])
names(chr19_SeqSet) <- "chr19"
writeXStringSet(chr19_SeqSet,filepath = "Data/chr19_Seq.fa")

```

```{r DNAStringSub61s1, echo=FALSE,eval=TRUE,collapse=FALSE}
chr19_SeqSet <- DNAStringSet(chr19_Seq[10000000:10000010])
names(chr19_SeqSet) <- "chr19"
writeXStringSet(chr19_SeqSet,filepath = "Data/chr19_Seq.fa")

```

---

#Reading a FASTA File

Now we can read our a FASTA file back to DNAStringSet object using the **readDNAStringSet()** function. 

We can access our DNAString from the DNAStringSet object again using **$** and **[[]]** functions.


```{r DNAStringSub62, echo=TRUE,eval=FALSE,tidy=TRUE}
chr19_FromFASTA <- readDNAStringSet(filepath = "Data/chr19_Seq.fa" )
chr19_FromFASTA$chr19
```

```{r DNAStringSubxzz62, echo=FALSE,eval=TRUE,tidy=TRUE}
chr19_FromFASTA <- readDNAStringSet(filepath = "Data/chr19_Seq.fa" )
chr19_FromFASTA$chr19
```


---

# Time for an exercise.

[Link_to_exercises](../../exercises/fastaAndBiostrings_exercises.html)

[Link_to_answers](../../answers/fastaAndBiostrings_answers.html)


