---
title: "Summarizing Scores In Bioconductor"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/SequencesInBioconductor/"
output: 
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```

```{r setup2, include=FALSE,eval=FALSE,echo=FALSE}
library(ShortRead)
temp <- readFastq("~/Projects/Results/RNAseqPipeTest/FirstTest/FQs/ENCFF000CXH.fastq.gz")
fastqSample <- temp[1:100000]
writeFastq(fastqSample,file = "~/Projects/Software/Github/RUBioconductor_Introduction/r_course/Data/sampled_ENCFF000CXH.fastq.gz",mode = "w")
```

```{r,results='asis',include=TRUE,echo=FALSE}
if(params$isSlides == "yes"){
  cat("---
class: inverse, center, middle

# Summarizing Scores

<html><div style='float:left'></div><hr color='#EB811B' size=1px width=720px></html> 

---
"    
  )
}else{
  cat("---

# Summarizing Scores

---
"    
  )
  
}
```


# Raw Sequence data

Sequencing data from Illumina sequences machines are often stored in FASTQ format.

<div align="center">
<img src="imgs/fastq.png" alt="igv" height="400" width="600">
</div>


---
# Aligned sequence data

Sequence reads aligned to a genome are most often stored in BAM format.

<div align="center">
<img src="imgs/sam.png" alt="igv" height="400" width="600">
</div>


---
# Summarizing Alignments

A common analysis step in the interpretation of high-throughput sequencing experiments is to quantify the abundance of sequence in/over genomic regions.

<div align="center">
<img src="imgs/sum.png" alt="igv" height="400" width="600">
</div>

---
# Summarizing for RNAseq

* **In RNAseq** experiments we assess the abundance of transcripts/genes under differing conditions by assessing the number of reads aligning with the transcript's/gene's exons.

![](imgs/RNAseq.png)
---
# Summarizing for ChIPseq

* **In ChIPseq** experiments we compare the number of DNA fragments mapping to a genomic locus after an IP for a specific protein or DNA/histone state versus the number of DNA fragments mapping from a background control/input sample.

![](imgs/chipseq.png)
---
# Summarizing Genomic Alignments
We can provide a summarization of aligned data per base pair using **wig, bedgraph and bigWig** files.

![](imgs/scoreBP.png)


---
# Summarizing alignments in/over regions

We will often summarise our aligned data in BAM format over genomic regions of interest.

![](imgs/countss.png)

# Summarizing alignments in/over regions

The genomic regions may be:

* Simple genomic regions such as peaks or TSS regions are often stored as **BED** files. In Bioconductor these are represented as **GRanges** objects.

* More complex, multi-part regions such as gene models are often stored as **GTF** or **GFF** files. In Bioconductor, gene models are handled as **TxDb** objects.

---
# Summarizing with Bioconductor

We will use two main functions to allows us to summarize our aligned data over the genome and within regions.

These are the **coverage()** and **summarizeOverlaps()** functions from **GenomicsAlignments** package.

* **coverage()** - To create per base pair signal/alignments scores across genome i.e. bigWigs.
* **summarizeOverlaps()** - To create counts of signal/alignments within genomic regions.

---
# The Data

In this session we will work with aligned data as BAM files. 

I have provided BAM files we saw in our last session as our test data from today.

This can be found in **Data/liver.bodyMap.bam** and **Data/heart.bodyMap.bam**.

* **Data/liver.bodyMap.bam**
* **Data/heart.bodyMap.bam**

---

# GenomicAlignments packages

To make use of a GenomicsAlignments package we must first load the library.

```{r load, echo=TRUE,eval=FALSE}
library(GenomicAlignments)
```

```{r load1, echo=FALSE,eval=TRUE,warning=FALSE}
suppressPackageStartupMessages(library(GenomicAlignments))
```

---
# Coverage

The **coverage()** function can accept a Bam file path and will return an [Rlelist](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html#22) of the number of reads at every base pair.

We should always work with sorted and indexed BAM files so we will perform these steps first.


```{r a1}
sortedHeart <- sortBam("../../Data/heart.bodyMap.bam","Heart")
indexBam(sortedHeart)

```

```{r a1232,include=FALSE}
sortedHeart <- sortBam("../../Data/liver.bodyMap.bam","Liver")
indexBam(sortedHeart)

```
---
# Coverage

Now we simply use the **coverage()** function with the path to our sorted file.

```{r a1.2}
heartCoverage <- coverage("Heart.bam")
class(heartCoverage)
```

---
# Coverage as RLElist

As we have seen,  [Rlelist](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html#22) objects contain compressed vectors carrying the scores, here number of alignments, for every base pair in the genome.

Here we see scores only on chromosome 12 around the gene we have been reviewing, SLC25A3.

```{r a2,eval=FALSE}
heartCoverage
```

```{r a22,echo=FALSE}
heartCoverage[3:4]
```

---
# Coverage as RLElist

We have seen we can  [index, subset, manipulate and plot our RleList and Rle vector](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html#24) in a similar manner to standard lists and vectors. 


```{r a3,fig.width=8,fig.height=2}
chr12Cov <- heartCoverage[["chr12"]]
signalDepth <- chr12Cov[98986183:98998558]
signalDepthScaled <- data.frame(Position=98986183:98998558,
                                Signal=signalDepth*1000)
library(ggplot2)
ggplot(signalDepthScaled,aes(x=Position,y=Signal))+
  geom_line()+theme_minimal()
```

---
# Coverage function.

The **coverage()** function will also accept 
**GAlignments** objects.

```{r a4}
heartAln <- readGAlignments("Heart.bam")
heartCov1 <- coverage(heartAln)
```

```{r frfga3,fig.width=8,fig.height=2,echo=FALSE}
chr12Cov <- heartCov1[["chr12"]]
signalDepth <- chr12Cov[98986183:98998558]
signalDepthScaled <- data.frame(Position=98986183:98998558,
                                Signal=signalDepth*1000)
library(ggplot2)
ggplot(signalDepthScaled,aes(x=Position,y=Signal))+
  geom_line()+theme_minimal()
```

---
# Coverage function.

As well as **GRanges** and **GRangesList** objects too.

For RNAseq we should generate coverage from a [**GRangesList** object to account for reads spannning exons](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/AlignedDataInBioconductor.html#36).

```{r a5}
heartGR <- granges(heartAln)
heartCov2 <- coverage(heartGR)
heartGRL <- grglist(heartAln)
heartCov3 <- coverage(heartGRL)
```

```{r frfa3,fig.width=8,fig.height=2,echo=FALSE}
chr12Cov <- heartCov3[["chr12"]]
signalDepth <- chr12Cov[98986183:98998558]
signalDepthScaled <- data.frame(Position=98986183:98998558,
                                Signal=signalDepth*1000)
library(ggplot2)
ggplot(signalDepthScaled,aes(x=Position,y=Signal))+
  geom_line()+theme_minimal()
```


---
# Coverage function.

This means we can perform some operations on our sequence reads in R prior to creating our bigWig.

Here we simply select only reads aligning to the positive strand prior to creating our signal scores..

```{r a6s,echo=TRUE,eval=FALSE}
heartAlnPos <- heartAln[strand(heartAln) == "+"]
heartAlnPos <- coverage(heartAlnPos)
heartAlnPos["chr12"]
export.bw(heartAlnPos,con="heartPos.bw")
```

```{r a6,echo=FALSE,eval=TRUE}
heartAlnPos <- heartAln[strand(heartAln) == "+"]
heartAlnPos <- coverage(heartAlnPos)
heartAlnPos["chr12"]
```




---
# Coverage function.

The coverage function allows for the specification of a additional parameter, **weight**.

The **weight** parameter applies a scaling factor to the calculated scores and so can be useful to normalise your signal , e.g. to total mapped reads.

```{r a7,collapse=TRUE}
heartCoverageX10 <- coverage("Heart.bam",
                          weight = 10)
heartCoverageX10["chr12"]
heartCoverage["chr12"]
```


---
# Coverage function.

In combination with **idxstatsBam()** function,  we could use this to normalise your signal to total mapped reads.

```{r a8}
allChromosomeStats <- idxstatsBam("Heart.bam")
allChromosomeStats
```

---
# Coverage function.


```{r a8k}
mapped <- sum(allChromosomeStats[,"mapped"])
heartCoverageNorm <- coverage("Heart.bam",
                          weight = (10^6)/mapped)
heartCoverageNorm["chr12"]
```

---
# Summarising from scores.

We can use [RleLists to summarise scores over regions by subsetting with GRanges and using standard arithmetic operations.](https://rockefelleruniversity.github.io/Bioconductor_Introduction/r_course/presentations/slides/GenomicScores_In_Bioconductor.html#36)

Here we get the mean signal (read depth per base pair) over exons for slc25A3 gene, **Entrez ID 5250**.

```{r a91,echo=FALSE}
suppressPackageStartupMessages(library(TxDb.Hsapiens.UCSC.hg19.knownGene))
```

```{r a9}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
exonsOfGenes <- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene,
                        by="gene")
slc25A3 <- exonsOfGenes[["5250"]]
slc25A3
```

---
# Summarising from scores 

```{r a9lk}
heartCoverageNorm <- coverage("Heart.bam")
myMeanCovOverExons <- mean(heartCoverageNorm[slc25A3])
myMeanCovOverExons
```

---
# Counting reads in/over regions

To count reads within regions and across gene models we can take advantage of specialist counting tools available within the **summariseOverlaps()** functions.

This will allow us to importantly

* Handle reads overlapping multiple regions (features or metafeatures).
* Assign reads to meta-features (genes instead of exons).

---
# Reads over/across exons.


![](imgs/exonsAndGenes.png)



<!-- --- -->
<!-- # SummariseOverlaps - Modes of counting -->

<!-- We can control counting behaviour by setting **mode**. Typically we will use the conservative mode **union**. -->

<!-- <div align="center"> -->
<!-- <img src="imgs/count_modes.png" alt="igv" height="400" width="400"> -->
<!-- </div> -->


---
# SummariseOverlaps - Summarisation to genes (not individual exons)

Often we will want to summarise to genes, not individual exons.

<div align="center">
<img src="imgs/withinGenes.png" alt="igv" height="400" width="600">
</div>

---
# SummariseOverlaps - A read only assigned to one feature

When using **summariseOverlaps()** function, a read can only be assigned to one feature.

This is important to note when counting features such as overlapping exons.

We will learn how to deal with counting across overlapping features in later sessions such as in RNA-seq.


---
# SummariseOverlaps

The **summariseOverlaps()** function is well intergrated with many of the Bioconductor objects we have seen.

As with the **coverage()** function it can count reads from **Bam** files or **GRanges** objects.

The **summariseOverlaps()** function can count over **GRanges (features e.g. exons)** and **GRangesLists (meta-features genes)**. 



---
# SummariseOverlaps for regions

To count non-overlapping regions, we can simply specify the path to our BAM file and a GRanges of regions to count over.

The returned object is a **RangedSummarizedExperiment** object.


```{r a11}

geneBody <- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
TSS <- promoters(geneBody,500,500)
myTSScounts <- summarizeOverlaps(TSS,"Heart.bam")
class(myTSScounts)
```

---
# SummarizedExperiment object

The resulting **SummarizedExperiment** object contains a matrix of our counts over regions and the original GRanges object we used for counting.

As with other Bioconductor objects, the object shows some useful summary information

```{r a11a}
myTSScounts
```

---
# SummarizedExperiment object

We can access the count matrix using the accessor function **assay()** with our **SummarizedExperiment** object.

Here we see total counts for SLC25A3 genes' TSS region


```{r a12,echo=TRUE,eval=FALSE}
countMatrix <- assay(myTSScounts)
countMatrix["5250",]
```

```{r a12s,echo=FALSE,eval=TRUE}
countMatrix <- assay(myTSScounts)
countMatrix["5250",,drop=FALSE]
```

---
# SummarizedExperiment object

The **rowRanges()** function will access the original **GRanges** we used for counting.

```{r a13,echo=TRUE}
Granges <- rowRanges(myTSScounts)
Granges
```


---
# SummariseOverlaps for gene models

For gene models, we may want to assign reads to each gene and so count all reads overlapping any exons within a gene.

We can count reads in meta-features such as genes using **GRangesLists**.

First we extract a GRangesList of exons for every gene.

```{r a131}

geneExons <- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene,by="gene")
geneExons["5250"]
```

---
# SummariseOverlaps for gene models

Now we can summarise counts for every gene by counting in their respective exons.

```{r a14a}
myGeneCounts <- summarizeOverlaps(geneExons,"Heart.bam")
myGeneCounts
```

---
# SummariseOverlaps for gene models

And now we have a count of signal in exons for every gene.


```{r a122a,echo=TRUE,eval=FALSE}
countMatrix <- assay(myGeneCounts)
countMatrix["5250",]
```

```{r a122b,echo=FALSE,eval=TRUE}
countMatrix <- assay(myGeneCounts)
countMatrix["5250",,drop=FALSE]
```

---
# SummariseOverlaps for gene models

The **rowRanges()**  contains the original GRangesList of exons by genes.

```{r a1211a}
grgList <- rowRanges(myGeneCounts)
grgList
```

---
# SummarizeOverlaps for many files.

The **summarizeOverlaps()** function can count across multiple BAM files.

We can specify a vector of BAM file paths we wish to count from to **summarizeOverlaps()** function.

```{r a1312a}
allGeneCounts <- summarizeOverlaps(geneExons,
                                   c("Heart.bam","Liver.bam"))
```

---
# SummarizeOverlaps for many files.

Now we can return the matrix of counts with samples by columns and genes as rows.


```{r a122ssa,echo=TRUE,eval=FALSE}
countMatrix <- assay(allGeneCounts)
countMatrix["5250",]
```

```{r a1dd22b,echo=FALSE,eval=TRUE}
countMatrix <- assay(allGeneCounts)
countMatrix["5250",,drop=FALSE]
```


---
# Low memory summarisation

Counting from BAM files, as with sorting, can be configured to trade of memory usage for speed.

i.e. less ram used but takes longer to count.

We can control the amount of memory used by creating a **BamFile** object prior to counting using the **BamFile()** function

```{r a20}
myBam <- BamFile("Heart.bam")
class(myBam)
```

---
# Low memory summarisation

We can additionally specify the **yieldSize** parameter which controls how many reads are held in memory in one time.

Here we set the **yieldSize**  to a very low 1000 reads, this will save a lot of memory but slow down processing.

We will often need to test the **yieldSize** which best fits our machine.

```{r a21q}
myBam <- BamFile("Heart.bam",yieldSize = 1000)
heartGeneCounts <- summarizeOverlaps(geneExons,myBam)
heartGeneCounts
```

---
# Low memory summarisation

When counting multiple files we can use the **BamFileList()** function in a similar manner but this time we provide a vector of the two BAM file paths.

```{r a22w}
myBam <- BamFileList(c("Heart.bam","Liver.bam"),
                     yieldSize = 1000)
allGeneCounts <- summarizeOverlaps(geneExons,myBam)
allGeneCounts
```

---
# Time for an exercise.

[Link_to_exercises](../../exercises/summarization_exercises.html)

[Link_to_answers](../../answers/summarization_answers.html)


