<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>GenomicFeatures_In_Bioconductor.utf8</title>
    <meta charset="utf-8" />
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="metropolisCustom.css" type="text/css" />
    <link rel="stylesheet" href="metropolis-fontsCustom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Genomic Features In Bioconductor
<html>
<div style="float:left">

</div>
<hr color='#EB811B' size=1px width=796px>
</html>
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Bioconductor_Introduction/" class="uri">http://rockefelleruniversity.github.io/Bioconductor_Introduction/</a>

---





---
class: inverse, center, middle

# Genomic Features

&lt;html&gt;&lt;div style='float:left'&gt;&lt;/div&gt;&lt;hr color='#EB811B' size=1px width=720px&gt;&lt;/html&gt; 

---



## Genomic Features

Genomic Features are often stored as GTF or GFF files.

&lt;div align="center"&gt;
&lt;img src="imgs/gff.png" alt="igv" height="200" width="600"&gt;
&lt;/div&gt;

---
## Genes as Genomic Features

In particular, we often find gene models as GFF or GTF files.

This importantly allows us to associate.

* Exons with transcripts.
* Transcripts with genes.

With this mapping we can perform complex operations such as summarising genome wide RNAseq signal to genes/transcripts expressions.  

---
## IGV and GFF/GTF

We use GFF and GTF in IGV to display gene models. 



&lt;div align="center"&gt;
&lt;img src="imgs/fullGenes.png" height="460" width="800"&gt;
&lt;/div&gt;

---
## A Single Gene Model in GTF Format

GTF is perhaps the most common format for gene models. 

In this example we have a GTF file containing a single gene model for the **Xkr4** gene.

This can be found in **data/Xkr4.gtf**

&lt;div align="center"&gt;
&lt;img src="imgs/gtfInTerminal.png" height="138" width="800"&gt;
&lt;/div&gt;
---
## A Single Gene Model in IGV

IGV parses information from the GTF file to construct the gene model we see containing untranslated regions and directionality of expression.

&lt;div align="center"&gt;
&lt;img src="imgs/gtfInIGV.png" height="500" width="800"&gt;
&lt;/div&gt;

---
## Genomic Features in Bioconductor

In Bioconductor, gene models are prepackaged for us in the **TxDb** packages.

Format is TxDb. **species** . **source** . **major version** . **table**

Homo Sapiens gene build from UCSC's version hg19 known gene table - TxDb.Hsapiens.UCSC.hg19.knownGene

&lt;div align="center"&gt;
&lt;img src="imgs/txdb.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;


---
## Genomic Features in Bioconductor

We can however make use of GTF/GFF files of our own gene models outside of Bioconductor pre-built TxDB packages.

The **GenomicFeatures** package has functions to import a GTF/GFF into a Bioconductor **TxDb** object as well as to interact with **TxDb** objects.

- [**GenomicFeatures**](https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html) - Import to and handling of **TxDB** objects in R.

---
## Genomic Features in Bioconductor

The first package we will look at is the **TxDB** package - **TxDb.Hsapiens.UCSC.hg19.knownGene**.

Remember we can install Bioconductor packages (fairly) easily  using Bioconductor package pages' provided commands. 


```r
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("TxDb.Hsapiens.UCSC.hg19.knownGene")
```

---
## Genomic Intervals in Bioconductor

Now we have the package installed, we can load the library.




```r
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

---
## TxDb

Once we have loaded the **TxDb.Hsapiens.UCSC.hg19.knownGene** library, we will have a new **TxDb** object available to us called **TxDb.Hsapiens.UCSC.hg19.knownGene**.


```r
class(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

```
## [1] "TxDb"
## attr(,"package")
## [1] "GenomicFeatures"
```
---
## TxDb

When this variable name is typed in into console we can retrieve some key information on the Gene Model build as well as summary metrics.


```r
TxDb.Hsapiens.UCSC.hg19.knownGene
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
```

---
## Accessing Information

As suggested in the name, **TxDb** objects are a database containing information we would store in GTF/GFF files.

To access information in our **TxDb** object we will need to use special accessor functions.

These accessors return information in **GRanges'** formats allowing us to work with methods from the **GenomicRanges** packages.

---
# Extracting Genomic Locations

We have a number of functions which allow us to extract GRanges for our features of interest including:

* **genes()** - Extract gene locations as a **GRanges**.
* **transcripts()** - Extract transcript locations  as a **GRanges**.
* **exons()** - Extract exon locations  as a **GRanges**.
* **cds()** - Extract coding locations  as a **GRanges**.
* **promoters()** - Extract promoter locations  as a **GRanges**.


---
## Extracting Genes

We can extract genes with the **genes()** function and the **TxDb** object as a parameter.


```r
myGenes &lt;- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
myGenes
```

```
## GRanges object with 23056 ranges and 1 metadata column:
##         seqnames              ranges strand |     gene_id
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19   58858172-58874214      - |           1
##      10     chr8   18248755-18258723      + |          10
##     100    chr20   43248163-43280376      - |         100
##    1000    chr18   25530930-25757445      - |        1000
##   10000     chr1 243651535-244006886      - |       10000
##     ...      ...                 ...    ... .         ...
##    9991     chr9 114979995-115095944      - |        9991
##    9992    chr21   35736323-35743440      + |        9992
##    9993    chr22   19023795-19109967      - |        9993
##    9994     chr6   90539619-90584155      + |        9994
##    9997    chr22   50961997-50964905      - |        9997
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
## Extracting Exons

We can extract exons with the **exons()** function and the **TxDb** object as a parameter in the same way.


```r
myExons &lt;- exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
myExons
```

```
## GRanges object with 289969 ranges and 1 metadata column:
##                  seqnames      ranges strand |   exon_id
##                     &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 11874-12227      + |         1
##        [2]           chr1 12595-12721      + |         2
##        [3]           chr1 12613-12721      + |         3
##        [4]           chr1 12646-12697      + |         4
##        [5]           chr1 13221-14409      + |         5
##        ...            ...         ...    ... .       ...
##   [289965] chrUn_gl000241 35706-35859      - |    289965
##   [289966] chrUn_gl000241 36711-36875      - |    289966
##   [289967] chrUn_gl000243 11501-11530      + |    289967
##   [289968] chrUn_gl000243 13608-13637      + |    289968
##   [289969] chrUn_gl000247   5787-5816      - |    289969
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
## Extracting Transcripts

We can extract transcripts with the **transcripts()** functions and the **TxDb** object as a parameter.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
myTranscripts
```

```
## GRanges object with 82960 ranges and 2 metadata columns:
##                 seqnames        ranges strand |     tx_id     tx_name
##                    &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##       [1]           chr1   11874-14409      + |         1  uc001aaa.3
##       [2]           chr1   11874-14409      + |         2  uc010nxq.1
##       [3]           chr1   11874-14409      + |         3  uc010nxr.1
##       [4]           chr1   69091-70008      + |         4  uc001aal.1
##       [5]           chr1 321084-321115      + |         5  uc001aaq.2
##       ...            ...           ...    ... .       ...         ...
##   [82956] chrUn_gl000237        1-2686      - |     82956  uc011mgu.1
##   [82957] chrUn_gl000241   20433-36875      - |     82957  uc011mgv.2
##   [82958] chrUn_gl000243   11501-11530      + |     82958  uc011mgw.1
##   [82959] chrUn_gl000243   13608-13637      + |     82959  uc022brq.1
##   [82960] chrUn_gl000247     5787-5816      - |     82960  uc022brr.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
## Extracting Coding Regions

We can extract coding regions with the **cds()** functions and the **TxDb** object as a parameter.


```r
myCDS &lt;- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)
myCDS
```

```
## GRanges object with 237533 ranges and 1 metadata column:
##                  seqnames        ranges strand |    cds_id
##                     &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1   12190-12227      + |         1
##        [2]           chr1   12595-12721      + |         2
##        [3]           chr1   13403-13639      + |         3
##        [4]           chr1   69091-70008      + |         4
##        [5]           chr1 324343-324345      + |         5
##        ...            ...           ...    ... .       ...
##   [237529] chrUn_gl000228   30530-31035      - |    237529
##   [237530] chrUn_gl000228   31353-31430      - |    237530
##   [237531] chrUn_gl000228   31660-31734      - |    237531
##   [237532] chrUn_gl000228   31660-31737      - |    237532
##   [237533] chrUn_gl000228   31996-32173      - |    237533
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Extracting Additional Information

We can control which additional information such as transcript ID and gene ID we retrieve in metadata columns using the **columns** parameter in the **exons()/transcripts()/cds()** functions. 

For a full list of available columsn see **?transcripts()** function help.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             columns=c("gene_id","tx_id"))
myTranscripts[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames      ranges strand |         gene_id     tx_id
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;integer&gt;
##   [1]     chr1 11874-14409      + |       100287102         1
##   [2]     chr1 11874-14409      + |       100287102         2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Extracting Promoter Regions

Another useful function is the **promoters()** function.

This accepts additional parameters of **upstream** and **downstream** to specify regions around TSS to be used as a promoter while accounting for a transcripts's strand.

We have been using the **resize()** function with **GRanges** objects for the same purpose.


```r
myPromoters &lt;- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             upstream=2000,downstream=50)
myPromoters[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##              seqnames     ranges strand |     tx_id     tx_name
##                 &lt;Rle&gt;  &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   uc001aaa.3     chr1 9874-11923      + |         1  uc001aaa.3
##   uc010nxq.1     chr1 9874-11923      + |         2  uc010nxq.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Extracting Groups of Genomic Features

A major component of gene models in GFF and GTF files is the grouping of exons to transcripts/genes and transcripts to genes.

We can extract lists of GRanges (a GRangesList) containing the grouping of features to metafeatures using additional accessor functions.

These include the **transcriptsBy()**, **exonsBy()** and **cdsBy()**.


---
## Extracting Groups of Genomic Features

The **transcriptsBy()** function takes as argument a **TxDb** object and a **by** argument specifying the metafeature/feature to group by.

For the **transcriptsBy()**, the **by** argument may be gene, exon or cds.


```r
transcriptByGenes &lt;- transcriptsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="gene")
transcriptByGenes[1:2]
```

```
## GRangesList object of length 2:
## $`1`
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`10`
## GRanges object with 1 range and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr8 18248755-18258723      + |     31944  uc003wyw.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Extracting Groups of Genomic Features

As with all lists we can extract information from elements in the GRangesList using the **$** and **[[]]** accessors.

This will extract the grouped intervals as a GRanges.


```r
# transcriptByGenes$1 or
transcriptByGenes[[1]]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames            ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 58858172-58864865      - |     70455  uc002qsd.4
##   [2]    chr19 58859832-58874214      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Extracting Groups of Genomic Features

Similarly the very useful **exonsBy()** function takes again an argument for a **TxDb** object and a **by** argument.

For the **exonsBy()** and **cdsBy()** functions, the **by** argument may be gene, or transcript.


```r
exonsByTranscript &lt;- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="tx")
exonsByTranscript[1:2]
```

```
## GRangesList object of length 2:
## $`1`
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames      ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]     chr1 11874-12227      + |         1        &lt;NA&gt;         1
##   [2]     chr1 12613-12721      + |         3        &lt;NA&gt;         2
##   [3]     chr1 13221-14409      + |         5        &lt;NA&gt;         3
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
## 
## $`2`
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames      ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;   &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]     chr1 11874-12227      + |         1        &lt;NA&gt;         1
##   [2]     chr1 12595-12721      + |         2        &lt;NA&gt;         2
##   [3]     chr1 13403-14409      + |         6        &lt;NA&gt;         3
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```


---
## Creating TxDb from External Sources

Often we will have our favourite gene models stored in a GTF outside of Bioconductor objects. If we wish to take advantage of all the features in Bioconductor we will need to import it into a **TxDb** object.

We can do this using the **makeTxDbFromGFF()** function which will parse GTF or GFF into a **TxDb** object

```r
myCustomTxDb &lt;- makeTxDbFromGFF("data/Xkr4.gtf")
```

```
## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
```

```r
class(myCustomTxDb)
```

```
## [1] "TxDb"
## attr(,"package")
## [1] "GenomicFeatures"
```

---
## Creating TxDb from External Sources

```r
myCustomTxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: data/Xkr4.gtf
## # Organism: NA
## # Taxonomy ID: NA
## # miRBase build ID: NA
## # Genome: NA
## # transcript_nrow: 1
## # exon_nrow: 3
## # cds_nrow: 3
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2020-04-30 19:10:12 -0400 (Thu, 30 Apr 2020)
## # GenomicFeatures version at creation time: 1.38.2
## # RSQLite version at creation time: 2.2.0
## # DBSCHEMAVERSION: 1.2
```

---
## Creating TxDb from External Sources

Now we have our own custom GTF object we can make use of the functions from **GenomicFeatures** package such as retrieving gene coordinates as a **GRanges**.


```r
genes(myCustomTxDb)
```

```
## GRanges object with 1 range and 1 metadata column:
##        seqnames          ranges strand |     gene_id
##           &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   Xkr4     chr1 3214482-3671498      - |        Xkr4
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
## Creating TxDb from External Sources

Or exons grouped by genes as a **GRangesList**.



```r
exonsBy(myCustomTxDb,by="gene")
```

```
## GRangesList object of length 1:
## $Xkr4
## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames          ranges strand |   exon_id   exon_name
##          &lt;Rle&gt;       &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr1 3214482-3216968      - |         1        &lt;NA&gt;
##   [2]     chr1 3421702-3421901      - |         2        &lt;NA&gt;
##   [3]     chr1 3670552-3671498      - |         3        &lt;NA&gt;
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
## Creating TxDb from UCSC

We can create a TxDb directly from UCSC's database using the **makeTxDbFromUCSC()** function.

The **makeTxDbFromUCSC** function simply requires a genome name to build a **TxDb** object. 

We can review the genomes available to us using the **ucscGenomes()** function from the **rtracklayer** package. This returns a data frame of genomes from which we can build a TxDB object.



```r
library(rtracklayer)
availableGenomes &lt;- ucscGenomes()
availableGenomes[1:4,]
```

```
##     db species      date                               name
## 1 hg38   Human Dec. 2013 Genome Reference Consortium GRCh38
## 2 hg19   Human Feb. 2009 Genome Reference Consortium GRCh37
## 3 hg18   Human Mar. 2006                    NCBI Build 36.1
## 4 hg17   Human  May 2004                      NCBI Build 35
```
---
## Creating TxDb from UCSC

Once we have identified the genome build we wish to use we can specify the genome parameter in the **makeTxDbFromUCSC()** function.


```r
hg18TxDb &lt;- makeTxDbFromUCSC(genome="hg18")
hg18TxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg18
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # UCSC Track: UCSC Genes
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: NA
## # transcript_nrow: 66803
## # exon_nrow: 266688
## # cds_nrow: 221991
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2020-04-30 19:10:52 -0400 (Thu, 30 Apr 2020)
## # GenomicFeatures version at creation time: 1.38.2
## # RSQLite version at creation time: 2.2.0
## # DBSCHEMAVERSION: 1.2
```
---
## Creating TxDb from UCSC

And again we can now make use of **GenomicFeatures** package's functions with our new TxDb objects


```r
hg18Promoters &lt;- promoters(hg18TxDb,2000,50)
```
---
## Exporting a TxDb Object as GTF/GFF File

We can make use of the **export.gff()** in the **rtracklayer** package to export our **TxDb** object to a gtf or gff file. We can include an additional argument of **format** to the **export.gff()** function.

We can specify the **format** to be **gff** or **gtf** as shown below.


```r
export.gff(myCustomTxDb,con="customTxDbb.gff",format="gff")
export.gff(myCustomTxDb,con="customTxDbb.gtf",format="gtf")
```
---
## Some Useful Functions in GenomicFeatures

We have seen that we can extract transcripts and exons grouped by their genes/transcripts using the **transcriptsBy()** and **exonsBy()** functions.

To identify the total number of groups we can simply use the **length()** function. Here we have 23459 transcipts or groups of exons.


```r
transcriptByGenes &lt;- exonsBy(hg18TxDb,by="gene")
length(transcriptByGenes)
```

```
## [1] 20121
```
---
## Some Useful Functions in GenomicFeatures

To find the number of exons in every group/transcript we can use the function **lengths()**. The **lengths()** function will tell us the length of every group in our **GRangesList**


```r
transcriptNumberPerGene &lt;- lengths(transcriptByGenes)
transcriptNumberPerGene[1:5]
```

```
##     1    10   100  1000 10000 
##    15     3    12    16    16
```
---
## Some Useful Functions in GenomicFeatures

We may wish to know the total sum length of all exons in every transcript. The **transcriptLengths()** function allows us to find all transcript lengths from a TxDb object as well as additional information.


```r
transcript_Lens &lt;- transcriptLengths(hg18TxDb)
transcript_Lens[1:5,]
```

```
##   tx_id    tx_name gene_id nexon tx_len
## 1     1 uc001aaa.2    &lt;NA&gt;     3   2122
## 2     2 uc009vip.1    &lt;NA&gt;     2   2772
## 3     3 uc009vjg.1    &lt;NA&gt;     3    708
## 4     4 uc009vjh.1   79501     3    888
## 5     5 uc001aal.1   79501     1    918
```
---
## Extracting Sequence Information

As with many Bioconductor packages, there is interoperability between **BSgenome** packages holding genome sequences and **TxDb** objects with gene models. One very useful function is the **extractTranscriptSeqs()** function.

The **extractTranscriptSeqs()** accepts a BSgenome object of genomic sequence of interest and a TxDb object.


```r
library(BSgenome.Hsapiens.UCSC.hg19)
hg19TransSeq &lt;- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, 
                                      transcripts=TxDb.Hsapiens.UCSC.hg19.knownGene)
hg19TransSeq
```

```
##   A DNAStringSet instance of length 82960
##          width seq                       names               
##     [1]   1652 CTTGCCGTCAG...GTTGGTTTCTG 1
##     [2]   1488 CTTGCCGTCAG...GTTGGTTTCTG 2
##     [3]   1595 CTTGCCGTCAG...GTTGGTTTCTG 3
##     [4]    918 ATGGTGACTGA...TAAAGTTTTAG 4
##     [5]     32 TACAGACCAAG...ACAATGGCCTA 5
##     ...    ... ...
## [82956]   1217 GCCAGTTTAGG...CTCCTGAGATC 82956
## [82957]    737 CTCCACTTCTG...TAGATGCAGTG 82957
## [82958]     30 TGGTGAATTTC...TGGCCAAAGAA 82958
## [82959]     30 TTGCAGAGGTG...CTCTTTGAGCC 82959
## [82960]     30 TGGTGAATTTC...TGGCCAAAGAA 82960
```
---
## Extracting Sequence Information

We can then either work with the resulting DNAStringSet then in R or write to Fasta file for analysis in external programs.


```r
writeXStringSet(hg19TransSeq,"myTranscriptSequences.fa")
```

---
## Changing Naming Conventions

So far we have used the standard UCSC style of chromosome names

- Chr1, Chr2, Chr3 ..

Sadly, conventions for chromosome and contig names can vary across different annotation sources.

Ensembl for instance uses.

- 1, 2, 3, 4 ...

---
## Changing Chromosome Naming Styles

The **GenomeInfoDb** package and associated **GenomeInfoDbData** package contain mappings between differing naming conventions and functions to convert between them.

First we need to load the **GenomeInfoDb** library.


```r
library(GenomeInfoDb)
```

---
## Reviewing Chromosome Naming Styles

We can review all the mappings for supported organisms' genomes using the **genomeStyles()** function with no arguments.

The **genomeStyles()** function returns a list of mapping data.frames of each organism.


```r
allMappings &lt;- genomeStyles()
names(allMappings)
```

```
##  [1] "Arabidopsis_thaliana"     "Caenorhabditis_elegans"   "Canis_familiaris"        
##  [4] "Cyanidioschyzon_merolae"  "Drosophila_melanogaster"  "Homo_sapiens"            
##  [7] "Mus_musculus"             "Oryza_sativa"             "Populus_trichocarpa"     
## [10] "Rattus_norvegicus"        "Saccharomyces_cerevisiae" "Zea_mays"
```

---
## Chromosome Naming Style

We can then review mappings for our organism of interest's genomes by retrieving the relevant entry from our list returned by the **genomeStyles()** function.

The resulting data.frame contains information on contig/chromosome mappings as well as additional information on chromosomes, i.e. Circularity. 


```r
#allMappings$Homo_sapiens or
allMappings[["Homo_sapiens"]]
```

```
##    circular  auto   sex NCBI  UCSC dbSNP Ensembl
## 1     FALSE  TRUE FALSE    1  chr1   ch1       1
## 2     FALSE  TRUE FALSE    2  chr2   ch2       2
## 3     FALSE  TRUE FALSE    3  chr3   ch3       3
## 4     FALSE  TRUE FALSE    4  chr4   ch4       4
## 5     FALSE  TRUE FALSE    5  chr5   ch5       5
## 6     FALSE  TRUE FALSE    6  chr6   ch6       6
## 7     FALSE  TRUE FALSE    7  chr7   ch7       7
## 8     FALSE  TRUE FALSE    8  chr8   ch8       8
## 9     FALSE  TRUE FALSE    9  chr9   ch9       9
## 10    FALSE  TRUE FALSE   10 chr10  ch10      10
## 11    FALSE  TRUE FALSE   11 chr11  ch11      11
## 12    FALSE  TRUE FALSE   12 chr12  ch12      12
## 13    FALSE  TRUE FALSE   13 chr13  ch13      13
## 14    FALSE  TRUE FALSE   14 chr14  ch14      14
## 15    FALSE  TRUE FALSE   15 chr15  ch15      15
## 16    FALSE  TRUE FALSE   16 chr16  ch16      16
## 17    FALSE  TRUE FALSE   17 chr17  ch17      17
## 18    FALSE  TRUE FALSE   18 chr18  ch18      18
## 19    FALSE  TRUE FALSE   19 chr19  ch19      19
## 20    FALSE  TRUE FALSE   20 chr20  ch20      20
## 21    FALSE  TRUE FALSE   21 chr21  ch21      21
## 22    FALSE  TRUE FALSE   22 chr22  ch22      22
## 23    FALSE FALSE  TRUE    X  chrX   chX       X
## 24    FALSE FALSE  TRUE    Y  chrY   chY       Y
## 25     TRUE FALSE FALSE   MT  chrM  chMT      MT
```

---
## Changing Chromosome Naming Styles

The **seqlevelsStyle()** function allows us to review as well as set the naming convention for our GenomicRanges or TxDb objects.

First lets review the genes from the  **TxDb.Hsapiens.UCSC.hg19.knownGene** package. Here we see they are **UCSC** style.


```r
myGenes &lt;- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
seqlevelsStyle(myGenes)
```

```
## [1] "UCSC"
```

```r
myGenes[1:2,]
```

```
## GRanges object with 2 ranges and 1 metadata column:
##      seqnames            ranges strand |     gene_id
##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##    1    chr19 58858172-58874214      - |           1
##   10     chr8 18248755-18258723      + |          10
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
## Changing Chromosome Naming Styles

Now we can update them to **Ensembl**, **NCBI** or **dbSNP** styles. Here we update to **Ensembl** chromosome naming conventions. 

This conversion is essential if we want to use our objects with data generated following Ensembl styles.


```r
seqlevelsStyle(myGenes) &lt;- "Ensembl"
myGenes[1:2,]
```

```
## GRanges object with 2 ranges and 1 metadata column:
##      seqnames            ranges strand |     gene_id
##         &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##    1       19 58858172-58874214      - |           1
##   10        8 18248755-18258723      + |          10
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```


---
## Gene Annotation

Where as the **TxDb** packages hold information on gene models such as exons positions and exon to gene mapping, the **OrgDb** packages contain information on gene's mapping to alternative IDs and any functional annotation. 

Information on model organism's gene annotation is contained with the **org.db** packages.

Format is org. **species** . **ID type** .db

Homo Sapiens annotation with Entrez Gene IDs -- org.Hs.eg.db

&lt;div align="center"&gt;
&lt;img src="imgs/orgdb.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;

---
## Gene Annotation

To gain access to gene annotation we must first load the library for the relevant species.

Here we install and load the library for Human, Org.Hs.eg.db.



```r
BiocManager::install("org.Hs.eg.db")
library(org.Hs.eg.db)
class(org.Hs.eg.db)
```


```
## [1] "OrgDb"
## attr(,"package")
## [1] "AnnotationDbi"
```

---
## Org.db AnnotationDbi Accessor Functions

As with the TxDb packages, we will need to use special accessors to retrieve information from Org.db.

We can use the functions:-

- **columns()** - Display what kind of annotation is available in OrgDb objects.
- **keytypes()** - Displays which type of identifiers can be used with **select** function. 
- **keys()** - Returns keys (index) for the database contained in the OrgDb object. Used along with **keytypes()** in &lt;b&gt;select&lt;/b&gt; function to retrieve interested annotation
- **select** - Retrieve the annotation data as a data.frame based on the supplied keys, keytypes and columns. 

We will explore how to retrieve annotation from gene-centric organism level annotation package (org.Hs.eg.db) using these functions.

---
## Accessing Annotation from Org.Db

We can use the **columns()** function first to list available annotation.

Here we see our annotation for alternative IDs (ENSEMBL, REFSEQ, GENENAME) as well as functional annotation (GO, PATH, OMIM)


```r
columns(org.Hs.eg.db)
```

```
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"       "UNIGENE"     
## [26] "UNIPROT"
```
--
To know more about the above identifier types

```r
help(GENENAME)
```
---
## Accessing Annotation from org.Hs.eg.db

Which keytypes can be used to query this database can be retrieved using keytypes.

```r
keytypes(org.Hs.eg.db)
```

```
##  [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
##  [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
## [11] "GO"           "GOALL"        "IPI"          "MAP"          "OMIM"        
## [16] "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"         "PMID"        
## [21] "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"       "UNIGENE"     
## [26] "UNIPROT"
```
---

## Accessing Annotation from Org.Db

If we want to extract few identifiers of a particular keytype, we can use keys() function

```r
keys(org.Hs.eg.db, keytype="SYMBOL")
```

```
##    [1] "A1BG"       "A2M"        "A2MP1"      "NAT1"       "NAT2"       "NATP"      
##    [7] "SERPINA3"   "AADAC"      "AAMP"       "AANAT"      "AARS1"      "AAVS1"     
##   [13] "ABAT"       "ABCA1"      "ABCA2"      "ABCA3"      "ABCB7"      "ABCF1"     
##   [19] "ABCA4"      "ABL1"       "AOC1"       "ABL2"       "ABO"        "ABR"       
##   [25] "ACAA1"      "ACACA"      "ACACB"      "ACADL"      "ACADM"      "ACADS"     
##   [31] "ACADSB"     "ACADVL"     "ACAT1"      "ACAT2"      "ASIC2"      "ASIC1"     
##   [37] "ACHE"       "ACLS"       "ACLY"       "ACO1"       "ACR"        "ACO2"      
##   [43] "ACOX1"      "ACP1"       "ACP2"       "ACP5"       "ACPP"       "ACRV1"     
##   [49] "ACTA1"      "ACTA2"      "ACTB"       "ACTBP2"     "ACTBP3"     "ACTBP4"    
##   [55] "ACTBP6"     "ACTBP7"     "ACTBP8"     "ACTBP9"     "ACTC1"      "ACTG1"     
##   [61] "ACTG2"      "ACTG1P1"    "ACTG1P2"    "ACTG1P3"    "ACTG1P6"    "ACTG1P7"   
##   [67] "ACTG1P8"    "ACTN4"      "ACTG1P9"    "ACTG1P10"   "ACTL6A"     "ACTN1"     
##   [73] "ACTN2"      "ACTN3"      "ACVR1"      "ACVR1B"     "ACVR2A"     "ACVR2B"    
##   [79] "ACVRL1"     "ACY1"       "ACYP1"      "ACYP2"      "ADA"        "ADAM8"     
##   [85] "ADAM10"     "ADAR"       "ADARB1"     "ADARB2"     "ADCP1"      "ADCY1"     
##   [91] "ADCY2"      "ADCY3"      "ADCY5"      "ADCY6"      "ADCY7"      "ADCY8"     
##   [97] "ADCY9"      "ADCYAP1"    "ADCYAP1R1"  "ADD1"       "ADD2"       "ADD3"      
##  [103] "ADFN"       "PLIN2"      "ADH1A"      "ADH1B"      "ADH1C"      "ADH4"      
##  [109] "ADH5"       "ADH6"       "ADH7"       "ADK"        "ADM"        "ADORA1"    
##  [115] "ADORA2A"    "ADORA2B"    "ADORA2BP1"  "ADORA3"     "ADPRH"      "PARP1"     
##  [121] "PARP4"      "PARP1P1"    "PARP1P2"    "ADRA1D"     "ADRA1B"     "ADRA1A"    
##  [127] "ADRA2A"     "ADRA2B"     "ADRA2C"     "ADRB1"      "ADRB2"      "ADRB3"     
##  [133] "GRK2"       "GRK3"       "ADSL"       "ADSS2"      "AP2A1"      "AP2A2"     
##  [139] "AP1B1"      "AP2B1"      "AP1G1"      "AEBP1"      "TLE5"       "CRISP1"    
##  [145] "AF8T"       "AFA"        "AFD1"       "AFG3L1P"    "AFM"        "AFP"       
##  [151] "AGA"        "ACAN"       "AGER"       "AGL"        "AGMX2"      "AGRP"      
##  [157] "JAG1"       "AGT"        "AGTR1"      "AGTR2"      "APLNR"      "AGXT"      
##  [163] "NR0B1"      "AHCY"       "AIC"        "AHR"        "AHSG"       "AIF1"      
##  [169] "AIH3"       "CRYBG1"     "AK1"        "AK2"        "AK4"        "AK4P1"     
##  [175] "AKT1"       "AKT2"       "ALAD"       "ALAS1"      "ALAS2"      "ALB"       
##  [181] "ALCAM"      "ABCD1"      "ALDH1A1"    "ALDH2"      "ALDH3A1"    "ALDH1B1"   
##  [187] "ALDH1A3"    "ALDH3B1"    "ALDH3B2"    "ALDH9A1"    "ALDH3A2"    "ABCD2"     
##  [193] "ALDOA"      "ALDOAP2"    "ALDOB"      "ALDOC"      "AKR1B1"     "AKR1B1P2"  
##  [199] "ALK"        "ALOX12"     "ALOX5"      "ALOX5AP"    "ALOX12B"    "ALOX12P1"  
##  [205] "ALOX12P2"   "ALOX15"     "ALOX15B"    "ALPI"       "ALPL"       "ALPP"      
##  [211] "ALPG"       "ALS3"       "ALX3"       "AMBN"       "AMBP"       "AMCN"      
##  [217] "AMD1"       "AMD1P2"     "AMELX"      "AMELY"      "AMFR"       "AMH"       
##  [223] "AMHR2"      "AMPD1"      "AMPD2"      "AMPD3"      "AMPH"       "BIN1"      
##  [229] "AMT"        "AMY1A"      "AMY1B"      "AMY1C"      "AMY2A"      "AMY2B"     
##  [235] "AMYP1"      "ANCR"       "ANG"        "ANGPT1"     "ANGPT2"     "ANK1"      
##  [241] "ANK2"       "ANK3"       "ANOP1"      "ANPEP"      "SLC25A4"    "SLC25A5"   
##  [247] "SLC25A6"    "ANXA1"      "ANXA2"      "ANXA2P1"    "ANXA2P2"    "ANXA2P3"   
##  [253] "ANXA3"      "ANXA4"      "ANXA5"      "ANXA6"      "ANXA7"      "ANXA11"    
##  [259] "ANXA13"     "AOAH"       "AOC2"       "AOX1"       "APAF1"      "NUDT2"     
##  [265] "APOF"       "APBA1"      "APBA2"      "APBB1"      "APBB2"      "APC"       
##  [271] "APCS"       "AIRE"       "APEH"       "APEX1"      "BIRC2"      "BIRC3"     
##  [277] "XIAP"       "BIRC5"      "APLP1"      "APLP2"      "APOA1"      "APOA2"     
##  [283] "APOA4"      "APOB"       "APOBEC1"    "APOC1"      "APOC1P1"    "AQP8"      
##  [289] "APOC2"      "APOC3"      "APOC4"      "APOD"       "APOE"       "APOH"      
##  [295] "APP"        "APRT"       "KLK3"       "FAS"        "FASLG"      "SHROOM2"   
##  [301] "AQP1"       "AQP2"       "AQP3"       "AQP4"       "AQP5"       "AQP6"      
##  [307] "AQP7"       "AQP9"       "AR"         "ABCC6"      "ARAF"       "ARCN1"     
##  [313] "TRIM23"     "AREG"       "ARF1"       "ARF3"       "ARF4"       "ARL4D"     
##  [319] "ARF4P1"     "ARF5"       "ARF6"       "ARG1"       "ARG2"       "RHOA"      
##  [325] "RHOB"       "RHOC"       "RND3"       "RHOG"       "ARHGAP1"    "ARHGAP4"   
##  [331] "ARHGAP5"    "ARHGAP6"    "ARHGDIA"    "ARHGDIB"    "ARHGDIG"    "RHOH"      
##  [337] "ARL1"       "PHOX2A"     "ARL2"       "ARL3"       "ARNT"       "ARNTL"     
##  [343] "ARR3"       "ARRB1"      "ARRB2"      "ARSA"       "ARSB"       "STS"       
##  [349] "ARSC2"      "ARSD"       "ARSL"       "ARSF"       "ART1"       "ART2P"     
##  [355] "ART3"       "ART4"       "ARVCF"      "ARVD3"      "ARVD4"      "ASAH1"     
##  [361] "ASCL1"      "ASCL2"      "ASD1"       "ASGR1"      "ASGR2"      "ASIP"      
##  [367] "ASL"        "ASLP1"      "ASMT"       "GET3"       "ASNS"       "ASNSP2"    
##  [373] "ASPA"       "ASPH"       "ASS1"       "ASS1P2"     "ASS1P4"     "ASS1P5"    
##  [379] "ASS1P6"     "ASS1P8"     "ASS1P9"     "ASS1P10"    "ASS1P14"    "ASTN1"     
##  [385] "SERPINC1"   "ZFHX3"      "ATD"        "ATF1"       "ATF3"       "ATF4"      
##  [391] "ATF4P1"     "ATHS"       "ATIC"       "ATM"        "RERE"       "ATOH1"     
##  [397] "ATOX1"      "ATP1A1"     "ATP1A2"     "ATP1A3"     "ATP12A"     "ATP1A4"    
##  [403] "ATP1B1"     "ATP1B2"     "ATP1B3"     "ATP1B3P1"   "ATP1B1P1"   "FXYD2"     
##  [409] "ATP2A1"     "ATP2A2"     "ATP2A3"     "ATP2B1"     "ATP2B2"     "ATP2B3"    
##  [415] "ATP2B4"     "ATP4A"      "ATP4B"      "ATP5F1A"    "ALDH7A1"    "ATP5F1AP2" 
##  [421] "ATP5F1B"    "ATP5F1BP1"  "ATP5F1C"    "ATP5F1D"    "ATP5F1E"    "ATP5PB"    
##  [427] "ATP5MC1"    "ATP5MC2"    "ATP5MC3"    "ATP5MC2P1"  "ATP5ME"     "ATP5PF"    
##  [433] "ATP6V1A"    "ATP6V1B1"   "ATP6V1B2"   "ATP6V0C"    "ATP6V1C1"   "ATP6V1E1"  
##  [439] "ATP6V0B"    "ATP6V1G2"   "ATP6V0A1"   "ATP6AP1"    "ATP7A"      "ATP5PO"    
##  [445] "ATP7B"      "ALDH7A1P1"  "ALDH7A1P2"  "ALDH7A1P3"  "ALDH7A1P4"  "ATR"       
##  [451] "ATRX"       "KIF1A"      "AUH"        "AUP1"       "AVP"        "AVPR1A"    
##  [457] "AVPR1B"     "AVPR2"      "AXL"        "AZF1"       "AZGP1"      "AZU1"      
##  [463] "B2M"        "B2MR"       "BAAT"       "BACH1"      "BAD"        "BAG1"      
##  [469] "BAGE"       "ADGRB1"     "ADGRB2"     "ADGRB3"     "BAK1"       "NKX3-2"    
##  [475] "BARD1"      "BAX"        "BBS1"       "BBS2"       "BBS4"       "BCAT1"     
##  [481] "BCAT2"      "BCHE"       "BCKDHA"     "BCKDHB"     "CCND1"      "BCL2"      
##  [487] "BCL2A1"     "BCL2L1"     "BCL2L2"     "BAK1P1"     "BAK1P2"     "BCL3"      
##  [493] "BCL5"       "BCL6"       "BCL7A"      "NBEAP1"     "BCL9"       "TNFRSF17"  
##  [499] "HCN2"       "OPN1SW"     "BCR"        "BCRP4"      "BCS1L"      "BCYRN1"    
##  [505] "BCYRN1P1"   "BCYRN1P2"   "BDH1"       "BDKRB1"     "BDKRB2"     "BDNF"      
##  [511] "BEVI"       "CFB"        "BFSP1"      "BGLAP"      "BGN"        "CEACAM1"   
##  [517] "BHMT"       "BICD1"      "BID"        "BIK"        "PRDM1"      "BLK"       
##  [523] "BLM"        "BLMH"       "CXCR5"      "BLVRA"      "BLVRB"      "BNC1"      
##  [529] "BMI1"       "BMP1"       "BMP2"       "BMP3"       "BMP4"       "BMP5"      
##  [535] "BMP6"       "BMP7"       "BMP8B"      "BMPR1A"     "BMPR1B"     "BMPR2"     
##  [541] "BMX"        "POLR3D"     "BNIP1"      "BNIP2"      "BNIP3"      "BNIP3L"    
##  [547] "BOK"        "DST"        "FOXL2"      "BPGM"       "BPHL"       "BPI"       
##  [553] "BRCA1"      "BRAF"       "BRCA2"      "BRDT"       "ZFP36L1"    "ZFP36L2"   
##  [559] "BRS3"       "BSG"        "BST1"       "BST2"       "BTC"        "BTD"       
##  [565] "KLF9"       "KLF5"       "BTF3"       "BTF3P11"    "BTF3P1"     "BTG1"      
##  [571] "BTK"        "BTN1A1"     "BTS1"       "BUB1"       "BVR1"       "BUB1B"     
##  [577] "BYSL"       "TSPO"       "C1QBP"      "SERPING1"   "ERC2-IT1"   "C1QA"      
##  [583] "C1QB"       "C1QC"       "C1R"        "C1S"        "C2"         "C3"        
##  [589] "C3AR1"      "C4A"        "C4B"        "C4BPA"      "C4BPAP2"    "C4BPB"     
##  [595] "CAPN5"      "C5"         "C5AR1"      "C6"         "C7"         "C8A"       
##  [601] "C8B"        "C8G"        "OSGIN2"     "C9"         "VPS51"      "MRPL49"    
##  [607] "ZNHIT2"     "MPPED2"     "MYRF"       "TMEM258"    "DAGLA"      "GAS8-AS1"  
##  [613] "FMNL1"      "LDLRAD4"    "PTTG1IP"    "CFAP410"    "TMEM50B"    "MPPED1"    
##  [619] "CA1"        "CA2"        "CA3"        "CA4"        "CA5A"       "CA5AP1"    
##  [625] "CA6"        "CA7"        "CA8"        "CA9"        "CA11"       "CA12"      
##  [631] "CACD"       "CACNA1A"    "CACNA1B"    "CACNA1C"    "CACNA1D"    "CACNA1E"   
##  [637] "CACNA1F"    "CACNA1S"    "DDR1"       "CACNA2D1"   "CACNB1"     "CACNB2"    
##  [643] "CACNB3"     "CACNB4"     "CACNG1"     "SLC25A20"   "SLC25A20P1" "CAD"       
##  [649] "CALB1"      "CALB2"      "S100G"      "CALCA"      "CALCB"      "CALCP"     
##  [655] "CALCR"      "CALD1"      "CALM1"      "CALM1P1"    "CALM1P2"    "CALM2"     
##  [661] "CALM2P2"    "CALM2P3"    "CALM3"      "CALML3"     "CALR"       "CALU"      
##  [667] "CAMK4"      "CAMK2A"     "CAMK2B"     "CAMK2D"     "CAMK2G"     "CAMLG"     
##  [673] "CAMP"       "CANX"       "CAPG"       "CAPN1"      "CAPN2"      "CAPN3"     
##  [679] "CAPNS1"     "CAPN6"      "CAPS"       "CAPZA1"     "CAPZA2"     "CAST"      
##  [685] "CAPZB"      "CARS1"      "CASP1"      "CASP2"      "CASP3"      "CASP4"     
##  [691] "CASP5"      "CASP6"      "CASP7"      "CASP8"      "CASP9"      "CASP10"    
##  [697] "CASQ1"      "CASQ2"      "CASR"       "CAT"        "CATR1"      "CAV1"      
##  [703] "CAV2"       "CAV3"       "RUNX2"      "RUNX1"      "RUNX1T1"    "CBFA2T3"   
##  [709] "RUNX3"      "CBFB"       "SERPINA6"   "CBL"        "CBLB"       "CBLN1"     
##  [715] "SERPINH1"   "CBR1"       "CBR3"       "CBS"        "CCA1"       "CCIN"      
##  [721] "CCAL1"      "KYAT1"      "CCK"        "CCKAR"      "CCKBR"      "KRIT1"     
##  [727] "CCNA2"      "CCNB1"      "CCNC"       "CCND2"      "CCND2P1"    "CCND3"     
##  [733] "CCND3P1"    "CCNE1"      "CCNF"       "CCNG1"      "CCNG2"      "CCNH"      
##  [739] "CCNT1"      "CCNT2"      "CCT"        "CCT6A"      "CD1A"       "CD1B"      
##  [745] "CD1C"       "CD1D"       "CD1E"       "CD2"        "CD3D"       "CD3E"      
##  [751] "CD3G"       "CD247"      "CD4"        "CD5"        "CD5L"       "CD6"       
##  [757] "CD7"        "CD8A"       "CD8B"       "CD8B2"      "CD9"        "CD14"      
##  [763] "CD19"       "MS4A1"      "MS4A3"      "CD22"       "CD24P1"     "CD24P2"    
##  [769] "CD24P3"     "CD24P4"     "CD27"       "CD28"       "CD80"       "CD86"      
##  [775] "TNFRSF8"    "TNFSF8"     "CD33"       "SIGLEC6"    "CD34"       "CD36"      
##  [781] "SCARB1"     "SCARB2"     "CD37"       "CD38"       "ENTPD1"     "ENTPD2"    
##  [787] "ENTPD6"     "ENTPD3"     "ENTPD5"     "CD40"       "CD40LG"     "CD44"      
##  [793] "CD47"       "CD48"       "CD53"       "CD58"       "CD59"       "CD63"      
##  [799] "CD68"       "CD69"       "CD70"       "CD72"       "CD74"       "CD79A"     
##  [805] "CD79B"      "CD81"       "ADGRE5"     "CD151"      "CDA"        "CDAN3"     
##  [811] "CDB2"       "CDK1"       "CDK11B"     "LRBA"       "CDC5L"      "SEPTIN7"   
##  [817] "CDC6"       "CDC20"      "CDC25A"     "CDC25B"     "CDC25C"     "CDC27"     
##  [823] "CDC34"      "CDC42"      "CDH1"       "CDH2"       "CDH3"       "CDH4"      
##  [829] "CDH5"       "CDH6"       "CDH7"       "CDH8"       "CDH9"       "CDH10"     
##  [835] "CDH11"      "CDH12"      "CDH12P2"    "CDH13"      "CDH15"      "CDH16"     
##  [841] "CDH17"      "CDH18"      "CDK2"       "CDK3"       "CDK4"       "CDK5"      
##  [847] "CDK6"       "CDK7"       "CDL2"       "CDK8"       "CDK9"       "CDKN1A"    
##  [853] "CDKN1B"     "CDKN1C"     "CDKN2A"     "CDKN2B"     "CDKN2C"     "CDKN2D"    
##  [859] "CDKN3"      "CDO1"       "CDR1"       "CDR2"       "CDS1"       "CDSN"      
##  [865] "CD52"       "CDX1"       "CDX2"       "CDX4"       "CLGN"       "CEACAM5"   
##  [871] "CEBPA"      "CEBPB"      "CEBPD"      "CEBPE"      "CEBPG"      "CECR"      
##  [877] "CEL"        "CELP"       "CENPA"      "CENPB"      "CENPC"      "CENPCP1"   
##  [883] "CENPE"      "CENPF"      "CES1"       "CETN1"      "CETN2"      "CETN3"     
##  [889] "CETP"       "CFL1"       "CFL2"       "CFL1P5"     "CTSC"       "CFTR"      
##  [895] "CGA"        "CGB3"       "CGF1"       "CEACAM3"    "CEACAMP5"   "CEACAMP11" 
##  [901] "CEACAM7"    "CEACAM8"    "CEACAM4"    "CEACAMP1"   "CEACAMP2"   "CEACAMP3"  
##  [907] "CEACAMP4"   "CEACAMP6"   "CEACAMP7"   "CEACAMP8"   "CEACAMP9"   "CEACAMP10" 
##  [913] "CHAD"       "RCBTB2"     "CHAT"       "RCC1"       "CHD1"       "CHD2"      
##  [919] "CHD3"       "CHD4"       "AKR1C4"     "CHEK1"      "FOXN3"      "CHGA"      
##  [925] "CHGB"       "CHI3L1"     "CHI3L2"     "CHIT1"      "CHKA"       "CHKB"      
##  [931] "CHM"        "CHML"       "CHN1"       "CHN2"       "CHRM1"      "CHRM2"     
##  [937] "LYST"       "CHRM3"      "CHRM4"      "CHRM5"      "CHRNA1"     "CHRNA2"    
##  [943] "CHRNA3"     "CHRNA4"     "CHRNA5"     "CHRNA7"     "CHRNB1"     "CHRNB2"    
##  [949] "CHRNB3"     "CHRNB4"     "CHRND"      "CHRNE"      "CHRNG"      "CHUK"      
##  [955] "CIDEA"      "CKB"        "CIRBP"      "CISH"       "TBCB"       "CKBE"      
##  [961] "CKBP1"      "CKM"        "CKMT1B"     "CKMT2"      "ERCC8"      "CKS1B"     
##  [967] "CKS2"       "CLA3"       "AP2M1"      "AP1S1"      "AP2S1"      "AP3S1"     
##  [973] "CLC"        "CLCA1"      "CLCN1"      "CLCN2"      "CLCN3"      "CLCN4"     
##  [979] "CLCN5"      "CLCN6"      "CLCN7"      "CLCNKA"     "CLCNKB"     "CLU"       
##  [985] "CLIC1"      "CLIC2"      "CLK1"       "CLK2"       "CLK2P1"     "CLK3"      
##  [991] "TPP1"       "CLN3"       "CLN5"       "CLNS1AP1"   "CLNS1A"     "CLPS"      
##  [997] "CLPTM1"     "CLTA"       "CLTB"       "CLTC"      
##  [ reached getOption("max.print") -- omitted 60164 entries ]
```
---
## Accessing Annotation from Org.Db

We can extract other annotations for a particular identifier using &lt;b&gt;select&lt;/b&gt; function

```r
select(org.Hs.eg.db, keys = "A1BG", keytype = "SYMBOL", 
       columns = c("SYMBOL", "GENENAME", "CHR") )
```

```
##   SYMBOL               GENENAME CHR
## 1   A1BG alpha-1-B glycoprotein  19
```

---
## Annotating Results

We can use this now to translate our Entrez IDs from TxDb object to a gene name

```r
geneLocations &lt;- genes(hg18TxDb)
geneLocations
```

```
## GRanges object with 19742 ranges and 1 metadata column:
##         seqnames              ranges strand |     gene_id
##            &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19   63549984-63565932      - |           1
##      10     chr8   18293035-18303003      + |          10
##     100    chr20   42681577-42713790      - |         100
##    1000    chr18   23784933-24011189      - |        1000
##   10000     chr1 241718158-242073207      - |       10000
##     ...      ...                 ...    ... .         ...
##    9991     chr9 114020538-114135733      - |        9991
##    9992    chr21   34658193-34665310      + |        9992
##    9993    chr22   17403795-17489967      - |        9993
##    9994     chr6   90596334-90640876      + |        9994
##    9997    chr22   49308863-49310900      - |        9997
##   -------
##   seqinfo: 49 sequences (1 circular) from hg18 genome
```

---
## Annotating Results


```r
IDs &lt;- geneLocations$gene_id
myTable &lt;- select(org.Hs.eg.db, keys = IDs, keytype = "ENTREZID",
                  columns = c("SYMBOL", "GENENAME", "ENTREZID") )
myTable[1:2,]
```

```
##   ENTREZID SYMBOL               GENENAME
## 1        1   A1BG alpha-1-B glycoprotein
## 2       10   NAT2  N-acetyltransferase 2
```
---

## Time for an Exercise

[Link_to_exercises](../../exercises/GenomicFeatures_exercises.html)

[Link_to_answers](../../answers/GenomicFeatures_answers.html)
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
