<!DOCTYPE html>
<html>
  <head>
    <title>Genomic Features In Bioconductor</title>
    <meta charset="utf-8">
    <meta name="author" content="Rockefeller University, Bioinformatics Resource Centre" />
    <link href="libs/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link href="libs/remark-css-0.0.1/default-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Genomic Features In Bioconductor
### Rockefeller University, Bioinformatics Resource Centre
### <a href="http://rockefelleruniversity.github.io/Bioconductor_Introduction/" class="uri">http://rockefelleruniversity.github.io/Bioconductor_Introduction/</a>

---




---

# Genomic Features

Genomic Features are often stored as GTF or GFF files.

&lt;div align="center"&gt;
&lt;img src="imgs/gff.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;

---
# Genes as Genomic Features

In particular, we often find gene models as GFF or GTF files.

This importantly allows us to associate.

* Exons with transcripts
* Transcripts with genes.

With this mapping we can perform complex operations such as summarising genome wide RNAseq signal to genes'/transcripts' expressions.  
---
# IGV and GFF/GTF.

We use GFF and GTF in IGV to display gene models. 



![](imgs/fullGenes.png)

---
# A single gene model in GTF format.

GTF is perhaps the most common format for gene models. 

In this example we have a GTF file containing a single gene model for the **Xkr4** gene.

This can be found in **Data/Xkr4.gtf**

![](imgs/gtfInTerminal.png)
---
# A single gene model in IGV.

IGV parses information from the GTF file to construct the gene model we see containing untranslated regions and directionality of expression.

![](imgs/gtfInIGV.png)

---
# Genomic Features in Bioconductor .

In Bioconductor, gene models are prepackaged for us in the **TxDb** packages.

Format is TxDb. **species** . **source** . **major version** . **table**

Homo Sapiens gene build from UCSC's version hg19 known gene table -- TxDb.Hsapiens.UCSC.hg19.knownGene

&lt;div align="center"&gt;
&lt;img src="imgs/txdb.png" alt="igv" height="400" width="400"&gt;
&lt;/div&gt;


---
# Genomic Features in Bioconductor.

We can however make use of GTF/GFF files of our own gene models outside of Bioconductor pre-built TxDB packages.

The GenomicFeatures package has functions to import a GTF/GFF into a Bioconductor TxDb object as well as to interact with TxDb objects.

- [**GenomicFeatures**](https://bioconductor.org/packages/release/bioc/html/GenomicFeatures.html) -- Import to and handling of TxDB objects in R.

---
# Genomic Features in Bioconductor.

The first package we will look at is the TxDB package - TxDb.Hsapiens.UCSC.hg19.knownGene.

Remember we can install Bioconductor packages (fairly) easily  using Bioconductor package pages' provided commands. 


```r
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite("TxDb.Hsapiens.UCSC.hg19.knownGene")
```

---
# Genomic Intervals in Bioconductor.

Now we have the package installed, we can load the library




```r
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

---
# TxDb

Once we have loaded the **TxDb.Hsapiens.UCSC.hg19.knownGene** library, we will have a new TxDb object available to us called **TxDb.Hsapiens.UCSC.hg19.knownGene** 


```r
class(TxDb.Hsapiens.UCSC.hg19.knownGene)
```

```
## [1] "TxDb"
## attr(,"package")
## [1] "GenomicFeatures"
```
---
# TxDb

When variable name is typed in into console we can retrieve some key information on the Gene Model build as well as summary metrics


```r
TxDb.Hsapiens.UCSC.hg19.knownGene
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg19
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: GRCh37
## # transcript_nrow: 82960
## # exon_nrow: 289969
## # cds_nrow: 237533
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2015-10-07 18:11:28 +0000 (Wed, 07 Oct 2015)
## # GenomicFeatures version at creation time: 1.21.30
## # RSQLite version at creation time: 1.0.0
## # DBSCHEMAVERSION: 1.1
```

---
# Accessing information.

As suggested in the name, **TxDb** objects are a database containing information we would store in GTF/GFF files.

To access information in our **TxDb** object we will need to use special accessor functions.

These accessors return information in **GRanges'** formats allowing us to work with methods from the **GenomicRanges** packages
---
# Extracting Genomic locations.

We have a number of functions which allow us to extract GRanges for our features of interest including

* genes() - Extract gene locations as a **GRanges**.
* transcripts() - Extract transcript locations  as a **GRanges**.
* exons() - Extract exon locations  as a **GRanges**.
* cds() - Extract coding locations  as a **GRanges**.
* promoters() - Extract promoter locations  as a **GRanges**.


---
# Extracting genes

We can extract genes with the **genes()** function and the **TxDb** object as a parameter.


```r
myGenes &lt;- genes(TxDb.Hsapiens.UCSC.hg19.knownGene)
myGenes
```

```
## GRanges object with 23056 ranges and 1 metadata column:
##         seqnames                 ranges strand |     gene_id
##            &lt;Rle&gt;              &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##       1    chr19 [ 58858172,  58874214]      - |           1
##      10     chr8 [ 18248755,  18258723]      + |          10
##     100    chr20 [ 43248163,  43280376]      - |         100
##    1000    chr18 [ 25530930,  25757445]      - |        1000
##   10000     chr1 [243651535, 244006886]      - |       10000
##     ...      ...                    ...    ... .         ...
##    9991     chr9 [114979995, 115095944]      - |        9991
##    9992    chr21 [ 35736323,  35743440]      + |        9992
##    9993    chr22 [ 19023795,  19109967]      - |        9993
##    9994     chr6 [ 90539619,  90584155]      + |        9994
##    9997    chr22 [ 50961997,  50964905]      - |        9997
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting exons

We can extract exons with the **exons()/transcripts()/cds()** functions and the **TxDb** object as a parameter in the same way.


```r
myExons &lt;- exons(TxDb.Hsapiens.UCSC.hg19.knownGene)
myExons
```

```
## GRanges object with 289969 ranges and 1 metadata column:
##                  seqnames         ranges strand |   exon_id
##                     &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 [11874, 12227]      + |         1
##        [2]           chr1 [12595, 12721]      + |         2
##        [3]           chr1 [12613, 12721]      + |         3
##        [4]           chr1 [12646, 12697]      + |         4
##        [5]           chr1 [13221, 14409]      + |         5
##        ...            ...            ...    ... .       ...
##   [289965] chrUn_gl000241 [35706, 35859]      - |    289965
##   [289966] chrUn_gl000241 [36711, 36875]      - |    289966
##   [289967] chrUn_gl000243 [11501, 11530]      + |    289967
##   [289968] chrUn_gl000243 [13608, 13637]      + |    289968
##   [289969] chrUn_gl000247 [ 5787,  5816]      - |    289969
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting transcripts

We can extract transcripts with the **transcripts()** functions and the **TxDb** object as a parameter.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
myTranscripts
```

```
## GRanges object with 82960 ranges and 2 metadata columns:
##                 seqnames           ranges strand |     tx_id     tx_name
##                    &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##       [1]           chr1 [ 11874,  14409]      + |         1  uc001aaa.3
##       [2]           chr1 [ 11874,  14409]      + |         2  uc010nxq.1
##       [3]           chr1 [ 11874,  14409]      + |         3  uc010nxr.1
##       [4]           chr1 [ 69091,  70008]      + |         4  uc001aal.1
##       [5]           chr1 [321084, 321115]      + |         5  uc001aaq.2
##       ...            ...              ...    ... .       ...         ...
##   [82956] chrUn_gl000237   [    1,  2686]      - |     82956  uc011mgu.1
##   [82957] chrUn_gl000241   [20433, 36875]      - |     82957  uc011mgv.2
##   [82958] chrUn_gl000243   [11501, 11530]      + |     82958  uc011mgw.1
##   [82959] chrUn_gl000243   [13608, 13637]      + |     82959  uc022brq.1
##   [82960] chrUn_gl000247   [ 5787,  5816]      - |     82960  uc022brr.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```
---
# Extracting Coding regions

We can extract coding regions with the **cds()** functions and the **TxDb** object as a parameter.


```r
myCDS &lt;- cds(TxDb.Hsapiens.UCSC.hg19.knownGene)
myCDS
```

```
## GRanges object with 237533 ranges and 1 metadata column:
##                  seqnames           ranges strand |    cds_id
##                     &lt;Rle&gt;        &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt;
##        [1]           chr1 [ 12190,  12227]      + |         1
##        [2]           chr1 [ 12595,  12721]      + |         2
##        [3]           chr1 [ 13403,  13639]      + |         3
##        [4]           chr1 [ 69091,  70008]      + |         4
##        [5]           chr1 [324343, 324345]      + |         5
##        ...            ...              ...    ... .       ...
##   [237529] chrUn_gl000228   [30530, 31035]      - |    237529
##   [237530] chrUn_gl000228   [31353, 31430]      - |    237530
##   [237531] chrUn_gl000228   [31660, 31734]      - |    237531
##   [237532] chrUn_gl000228   [31660, 31737]      - |    237532
##   [237533] chrUn_gl000228   [31996, 32173]      - |    237533
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting additional information for exons/transcripts/cds regions.

We can control which additional information such as transcript ID and gene ID we retrieve in metadata columns using the **columns** parameter in the **exons()/transcripts()/cds()** functions. 

For a full list of available columsn see **?transcripts()** function help.


```r
myTranscripts &lt;- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             columns=c("gene_id","tx_id"))
myTranscripts[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames         ranges strand |         gene_id     tx_id
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;CharacterList&gt; &lt;integer&gt;
##   [1]     chr1 [11874, 14409]      + |       100287102         1
##   [2]     chr1 [11874, 14409]      + |       100287102         2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting Promoter regions

Another useful function is the **promoters()** function.

This accepts additional parameter of **upstream** and **downstream** to specify regions around TSS to be used as a promoter accounting for a transcripts's strand.

We have been using the **resize()** function with **GRanges** objects for the same purpose.


```r
myPromoters &lt;- promoters(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                             upstream=2000,downstream=50)
myPromoters[1:2]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames        ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr1 [9874, 11923]      + |         1  uc001aaa.3
##   [2]     chr1 [9874, 11923]      + |         2  uc010nxq.1
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

A major component of gene models in GFF and GTF files is the grouping of exons to transcripts/genes and transcripts to genes.

We can extract lists of GRanges (a GRangesList) containing the grouping of features to metafeatures using additional accessor functions.

These include the **transcriptsBy()**, **exonsBy()** and **cdsBy()**


---
# Extracting groups of genomic features.

The **transcriptsBy()** function takes as argument a **TxDb** object and a **by** argument specifying the metafeature/feature to group by.

For the **transcriptsBy()**, the **by** argument may be gene, exon or cds


```r
transcriptByGenes &lt;- transcriptsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="gene")
transcriptByGenes[1:2]
```

```
## GRangesList object of length 2:
## $1 
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames               ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 [58858172, 58864865]      - |     70455  uc002qsd.4
##   [2]    chr19 [58859832, 58874214]      - |     70456  uc002qsf.2
## 
## $10 
## GRanges object with 1 range and 2 metadata columns:
##       seqnames               ranges strand | tx_id    tx_name
##   [1]     chr8 [18248755, 18258723]      + | 31944 uc003wyw.1
## 
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

As with all lists we can extract information from elements in the GRangesList using the **$** and **[[]]** accessors.

This will extract the grouped intervals as a GRanges.


```r
# transcriptByGenes$1 or
transcriptByGenes[[1]]
```

```
## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames               ranges strand |     tx_id     tx_name
##          &lt;Rle&gt;            &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]    chr19 [58858172, 58864865]      - |     70455  uc002qsd.4
##   [2]    chr19 [58859832, 58874214]      - |     70456  uc002qsf.2
##   -------
##   seqinfo: 93 sequences (1 circular) from hg19 genome
```

---
# Extracting groups of genomic features.

Similarly the very useful **exonsBy()** function takes again an argument for a **TxDb** object and a **by** argument.

For the **exonsBy()** and **cdsBy()** functions, the **by** argument may be gene, or tx


```r
exonsByTranscript &lt;- exonsBy(TxDb.Hsapiens.UCSC.hg19.knownGene, 
                                   by="tx")
exonsByTranscript[1:2]
```

```
## GRangesList object of length 2:
## $1 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand |   exon_id   exon_name exon_rank
##          &lt;Rle&gt;      &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt; &lt;integer&gt;
##   [1]     chr1 [11874, 12227]      + |         1        &lt;NA&gt;         1
##   [2]     chr1 [12613, 12721]      + |         3        &lt;NA&gt;         2
##   [3]     chr1 [13221, 14409]      + |         5        &lt;NA&gt;         3
## 
## $2 
## GRanges object with 3 ranges and 3 metadata columns:
##       seqnames         ranges strand | exon_id exon_name exon_rank
##   [1]     chr1 [11874, 12227]      + |       1      &lt;NA&gt;         1
##   [2]     chr1 [12595, 12721]      + |       2      &lt;NA&gt;         2
##   [3]     chr1 [13403, 14409]      + |       6      &lt;NA&gt;         3
## 
## -------
## seqinfo: 93 sequences (1 circular) from hg19 genome
```


---
# Creating TxDb from external sources.

Often we will have our favourite gene models stored in a GTF outside of Bioconductor objects. If we wish to take advantage of all the features in Bioconductor we will need to import it into a TxDb object.

We can do this using the **makeTxDbFromGFF()** function which will parse GTF or GFF into a TxDb object

```r
myCustomTxDb &lt;- makeTxDbFromGFF("../../Data/Xkr4.gtf")
```

```
## Import genomic features from the file as a GRanges object ... OK
## Prepare the 'metadata' data frame ... OK
## Make the TxDb object ... OK
```

```r
myCustomTxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: ../../Data/Xkr4.gtf
## # Organism: NA
## # Taxonomy ID: NA
## # miRBase build ID: NA
## # Genome: NA
## # transcript_nrow: 1
## # exon_nrow: 3
## # cds_nrow: 3
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2018-04-13 07:22:29 -0400 (Fri, 13 Apr 2018)
## # GenomicFeatures version at creation time: 1.28.4
## # RSQLite version at creation time: 2.0
## # DBSCHEMAVERSION: 1.1
```

---
# Creating TxDb from external sources.

Now we have our own custom GTF object we can make use of the functions from **GenomicFeatures** package such as retrieving gene coordinates as a **GRanges**.


```r
genes(myCustomTxDb)
```

```
## GRanges object with 1 range and 1 metadata column:
##        seqnames             ranges strand |     gene_id
##           &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   Xkr4     chr1 [3214482, 3671498]      - |        Xkr4
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
# Creating TxDb from external sources.

Or exons grouped by genes as a **GRangesList**



```r
exonsBy(myCustomTxDb,by="gene")
```

```
## GRangesList object of length 1:
## $Xkr4 
## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames             ranges strand |   exon_id   exon_name
##          &lt;Rle&gt;          &lt;IRanges&gt;  &lt;Rle&gt; | &lt;integer&gt; &lt;character&gt;
##   [1]     chr1 [3214482, 3216968]      - |         1        &lt;NA&gt;
##   [2]     chr1 [3421702, 3421901]      - |         2        &lt;NA&gt;
##   [3]     chr1 [3670552, 3671498]      - |         3        &lt;NA&gt;
## 
## -------
## seqinfo: 1 sequence from an unspecified genome; no seqlengths
```
---
# Creating TxDb from UCSC.

We can create a TxDb directly from UCSC's database using the **makeTxDbFromUCSC()** function.

The **makeTxDbFromUCSC** function simply requires a genome name to build a **TxDb** object. 

We can review the genomes available to us using the **ucscGenomes()** function from the **rtracklayer** package. This returns a data frame of genomes from which we can build a TxDB object.



```r
library(rtracklayer)
availableGenomes &lt;- ucscGenomes()
availableGenomes[1:4,]
```

```
##     db species      date                               name
## 1 hg38   Human Dec. 2013 Genome Reference Consortium GRCh38
## 2 hg19   Human Feb. 2009 Genome Reference Consortium GRCh37
## 3 hg18   Human Mar. 2006                    NCBI Build 36.1
## 4 hg17   Human  May 2004                      NCBI Build 35
```
---
# Creating TxDb from UCSC.

Once we have identified the genome build we wish to use we can specify the genome parameter in the **makeTxDbFromUCSC()** function.


```r
hg18TxDb &lt;- makeTxDbFromUCSC(genome="hg18")
hg18TxDb
```

```
## TxDb object:
## # Db type: TxDb
## # Supporting package: GenomicFeatures
## # Data source: UCSC
## # Genome: hg18
## # Organism: Homo sapiens
## # Taxonomy ID: 9606
## # UCSC Table: knownGene
## # UCSC Track: UCSC Genes
## # Resource URL: http://genome.ucsc.edu/
## # Type of Gene ID: Entrez Gene ID
## # Full dataset: yes
## # miRBase build ID: NA
## # transcript_nrow: 66803
## # exon_nrow: 266688
## # cds_nrow: 221991
## # Db created by: GenomicFeatures package from Bioconductor
## # Creation time: 2018-04-12 19:25:52 -0400 (Thu, 12 Apr 2018)
## # GenomicFeatures version at creation time: 1.28.4
## # RSQLite version at creation time: 2.0
## # DBSCHEMAVERSION: 1.1
```
---
# Creating TxDb from UCSC.

And again we can now make use of **GenomicFeatures** package's functions with our new TxDb objects


```r
hg18Promoters &lt;- promoters(hg18TxDb,2000,50)
```
---
# Exporting a TxDb object as a GTF/GFF.

We can make use of the **export.gff()** in the **rtracklayer** package to export our **TxDb** object to a gtf or gff file. We can include an additional argument of **format** to the **export.gff()** function.

We can specify the **format** to be **gff** or **gtf** as shown below.


```r
export.gff(myCustomTxDb,con="customTxDbb.gff",format="gff")
export.gff(myCustomTxDb,con="customTxDbb.gff",format="gtf")
```
---
# Some useful functions for working with gene model in GenomicFeatures.

We have seen that we can extract transcripts and exons grouped by their genes/transcripts using the **transcriptsBy()** and **exonsBy()** functions.

To identify the total number of groups we can simply use the **length()** function. Here we have 23459 transcipts or groups of exons.


```r
transcriptByGenes &lt;- exonsBy(hg18TxDb,by="gene")
length(transcriptByGenes)
```
---
# Some useful functions for working with gene model in GenomicFeatures.

To find the number of exons in every group/transcript we can use the function **lengths()**. The **lengths()** function will tell us the length of every group in our **GRangesList**


```r
transcriptNumberPerGene &lt;- lengths(transcriptByGenes)
transcriptNumberPerGene[1:5]
```

```
##     1    10   100  1000 10000 
##     2     1     1     2     3
```
---
# Some useful functions for working with gene model in GenomicFeatures.

We may wish to know the total sum length of all exons in every transcript. The **transcriptLengths()** function allows us to find all transcript lengths from a TxDb object as well as additional information.


```r
transcript_Lens &lt;- transcriptLengths(hg18TxDb)
transcript_Lens[1:5,]
```

```
##   tx_id    tx_name gene_id nexon tx_len
## 1     1 uc001aaa.2    &lt;NA&gt;     3   2122
## 2     2 uc009vip.1    &lt;NA&gt;     2   2772
## 3     3 uc009vjg.1    &lt;NA&gt;     3    708
## 4     4 uc009vjh.1   79501     3    888
## 5     5 uc001aal.1   79501     1    918
```
---
# Extracting sequence information.

As with many Bioconductor packages, there is interoperability between **BSgenome** packages holding genome sequences and **TxDb** objects with gene models. One very useful function is the **extractTranscriptSeqs()** function.

The **extractTranscriptSeqs()** accepts a BSgenome object of genomic sequence of interest and a TxDb object.


```r
library(BSgenome.Hsapiens.UCSC.hg19)
hg19TransSeq &lt;- extractTranscriptSeqs(BSgenome.Hsapiens.UCSC.hg19, 
                                      transcripts=TxDb.Hsapiens.UCSC.hg19.knownGene)
hg19TransSeq
```

```
##   A DNAStringSet instance of length 82960
##          width seq                                     names               
##     [1]   1652 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 1
##     [2]   1488 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 2
##     [3]   1595 CTTGCCGTCAGCCTTTTC...GCACACTGTTGGTTTCTG 3
##     [4]    918 ATGGTGACTGAATTCATT...TCTAGTGTAAAGTTTTAG 4
##     [5]     32 TACAGACCAAGCTCATGACTCACAATGGCCTA        5
##     ...    ... ...
## [82956]   1217 GCCAGTTTAGGGTCTCTG...CCTCCGCCTCCTGAGATC 82956
## [82957]    737 CTCCACTTCTGATCCTCC...GCTTTATTAGATGCAGTG 82957
## [82958]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA          82958
## [82959]     30 TTGCAGAGGTGGCTGGTTGCTCTTTGAGCC          82959
## [82960]     30 TGGTGAATTTCAGCCAAAGTGGCCAAAGAA          82960
```
---
# Extracting sequence information.

We can then either work with the resulting DNAStringSet then in R or write to Fasta file for analysis in external programs.


```r
writeXStringSet(hg19TransSeq,"myTranscriptSequences.fa")
```
    </textarea>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function() {
  var d = document, s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})();</script>

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  tex2jax: {
    skipTags: ['script', 'noscript', 'style', 'textarea', 'pre']
  }
});
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
