---
title: "Sequences In Bioconductor"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "http://rockefelleruniversity.github.io/SequencesInBioconductor/"
output: 
  xaringan::moon_reader:
    lib_dir: libs
    nature:
      highlightStyle: github
      highlightLines: true
      countIncrementalSlides: false
  html_document:
    toc: true # table of content true
    toc_float: yes
    depth: 3  # upto three depths of headings (specified by #, ## and ###)
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style


---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
AsSlides <- TRUE
```


# Genomic Sequences

As we have seen earlier, genomic sequences are often stored in FASTA format.

<div align="center">
<img src="imgs/fasta.png" alt="igv" height="400" width="400">
</div>


---

# Genomic Sequences in Bioconductor.

Genomic sequences can be handled in Bioconductor using the functions in the **Biostrings** package and full genome sequences for many model organisms are contained within the **BSgenome** packages.

---

# BSgenome packages.

To make use of a BSgenome package we must first install and load the library.

The resulting BSgenome object (here BSgenome.Mmusculus.UCSC.mm10) contains the full genome sequence for the model organism.

```{r load, echo=TRUE,eval=FALSE}
library(BSgenome.Mmusculus.UCSC.mm10)
class(BSgenome.Mmusculus.UCSC.mm10)
```

```{r load1, echo=FALSE,eval=TRUE}
suppressPackageStartupMessages(library(BSgenome.Mmusculus.UCSC.mm10))
class(BSgenome.Mmusculus.UCSC.mm10)
```
---

```{r theObject, echo=TRUE,eval=TRUE,collapse=FALSE}
BSgenome.Mmusculus.UCSC.mm10
```
---
# Data in BSgenome objects.

Simply by typing object in we have retrieved some important information.

Importantly it tells us how we can access information in the object.

- **seqnames()** function to retrieve all contig names.
- **$** and **[[]]** accessors to retrieve a given contig's sequence.
---
# Contig names.

We can retrieve contig names using **seqnames()** function.

```{r contignames, echo=TRUE,eval=TRUE,collapse=FALSE}
contigNames <- seqnames(BSgenome.Mmusculus.UCSC.mm10)
contigNames[1:22]
```
---

# Contig lengths

We can also retrieve contig lengths using **seqlengths()** function.

```{r contiglengths, echo=TRUE,eval=TRUE,collapse=FALSE}
contigLengths <- seqlengths(BSgenome.Mmusculus.UCSC.mm10)
contigLengths[1:4]
```

---

# Retreiving sequence information.

Now we know sequence names, we can retrieve sequence information using either the **$** or **[[]]** accessors.

Here we use the data.frame and list accessor **$** with name of contig of interest
```{r access, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq <- BSgenome.Mmusculus.UCSC.mm10$chr19
chr19_Seq
```
---
# Retreiving sequence information.


now we use list accessor **[[]]** with name of contig of interest
```{r access2, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq <- BSgenome.Mmusculus.UCSC.mm10[["chr19"]]
chr19_Seq
```

---

# Biostrings and DNAstring

The returned sequence object is a **DNAstring** object from the **Biostrings** package. DNAstrings are efficient objects for storing and accessing large sequences.

```{r dnastring, echo=TRUE,eval=TRUE,collapse=FALSE}
class(chr19_Seq)
```

**?DNAstring** for full help on DNAstring objects.
---

# Subsetting a DNAstring set 

We can subset a DNAstring object just like indexing a normal vector

```{r dnastringSub, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Seq[1:10000000]
```

---

# Subsetting a DNAstring set 

Or we can use the Biostrings function **subseq()** on our DNAstring.

Note that **subseq()** is a faster implementation of subsetting for DNAstring objects.

```{r dnastringSub2, echo=TRUE,eval=TRUE,collapse=FALSE}
subseq(chr19_Seq,start=1,end=10000000)
```

---

# Useful functions.

The Biostrings package has many useful functions for handling sequence data.

We can review the frequency of bases using the **alphabetFrequency()** function.

```{r dnastringSub3, echo=TRUE,eval=TRUE,collapse=FALSE}
alphabetFrequency(chr19_Seq)
```

---

# Useful functions.

We can get the complement,reverse and reverse complement of sequence using the **complement()**, **reverse()**, **reverseComplement()**.

```{r dnastringSub4, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_SeqComp <- complement(chr19_Seq)
chr19_SeqRev <- reverse(chr19_Seq)
chr19_SeqRevComp <- reverseComplement(chr19_Seq[10000000:10000010])
chr19_Seq[10000000:10000010]
chr19_SeqRevComp
```

---

# Useful functions.

We can get also match and count the occurrence of patterns in our sequence using **matchPattern()** and **countPattern()**.

We will look more in depth at these later in the course.

```{r dnastringSub5, echo=TRUE,eval=TRUE,collapse=FALSE}
chr19_Count <- countPattern(pattern="ATCTGCAATG",chr19_Seq)
chr19_Count
```

---

#Writing to a FASTA file.

The **Biostrings** package contains useful functions to read and write to a FASTA file.

To write our DNAString object to a FASTA file we can use the **writeXStringSet()** function. 

First we will convert our DNAString to a DNAStringSet object using the **DNAStringSet()** function and set names using the **names()** function.


```{r dnastringSub611, echo=TRUE,eval=FALSE,collapse=FALSE}
chr19_SeqSet <- DNAStringSet(chr19_Seq[10000000:10000010])
names(chr19_SeqSet) <- "chr19"
writeXStringSet(chr19_SeqSet,filepath = "Data/chr19_Seq.fa")

```

```{r dnastringSub61s1, echo=FALSE,eval=TRUE,collapse=FALSE}
chr19_SeqSet <- DNAStringSet(chr19_Seq[10000000:10000010])
names(chr19_SeqSet) <- "chr19"
writeXStringSet(chr19_SeqSet,filepath = "../../Data/chr19_Seq.fa")

```

---

#Reading a FASTA file.

Now we can read our a FASTA file back to DNAStringSet object using the **readDNAStringSet()** function. 

We can access our DNAString from the DNAStringSet object again using **$** and **[[]]** functions.


```{r dnastringSub62, echo=TRUE,eval=FALSE,tidy=TRUE}
chr19_FromFASTA <- readDNAStringSet(filepath = "Data/chr19_Seq.fa" )
chr19_FromFASTA$chr19
```

```{r dnastringSubxzz62, echo=FALSE,eval=TRUE,tidy=TRUE}
chr19_FromFASTA <- readDNAStringSet(filepath = "../../Data/chr19_Seq.fa" )
chr19_FromFASTA$chr19
```


---

# Time for an exercise.

[Link_to_exercises](../../exercises/fastaAndBiostrings_exercises.html)

[Link_to_answers](../../answers/fastaAndBiostrings_answers.html)


