---
title: "FASTA and BioString exercises"
author: "Rockefeller University, Bioinformatics Resource Centre"
date: "https://rockefelleruniversity.github.io/Bioconductor_Introduction/"
output: 
  html_document:
    number_sections: false  ## if you want number sections at each table header
    theme: united  # many options for theme, this one is my favorite.
    highlight: tango  # specifies the syntax highlighting style
params:
  toMessage: false    
---

```{r, setup,include=FALSE}
toMessage <- ifelse(!is.null(params$toMessage),params$toMessage,FALSE)
message(toMessage)
```


## Working with BSgenome objects using Biostrings

In these exercises will gain some experience working with the BSgenome packages.



1. Load the library BSgenome.Hsapiens.UCSC.hg19

```{r installBS,  echo=toMessage}
suppressPackageStartupMessages(library(BSgenome.Hsapiens.UCSC.hg19))
```

```{r installBSs,  echo=toMessage}
library(BSgenome.Hsapiens.UCSC.hg19)
```

2. Count the number of contigs.

```{r installBS2,  echo=toMessage}
length(seqnames(BSgenome.Hsapiens.UCSC.hg19))
```

3. Give the sum of lengths of the 3 smallest chromosomes.

```{r installBS2dd3,  echo=toMessage}
contigLengths <- seqlengths(BSgenome.Hsapiens.UCSC.hg19)
sum(contigLengths[order(contigLengths,decreasing = FALSE)][1:3])
```

4. How many unknown bases - base *N* -  are in chromosome 20

```{r installBS23,  echo=toMessage}
alpFreq <-alphabetFrequency(BSgenome.Hsapiens.UCSC.hg19$chr20)
alpFreq["N"]
```

5. Create a barchart of the total number of the A,T,C,G bases on chromosome 20.

```{r installBSs23,  echo=toMessage}
library(ggplot2)
alpFreq <-alphabetFrequency(BSgenome.Hsapiens.UCSC.hg19$chr20)
atcgFreq <- alpFreq[c("A","T","C","G")]
atcgFreqDF <- data.frame(Bases=names(atcgFreq),Frequency=atcgFreq)
ggplot(atcgFreqDF,aes(y=Frequency,x=Bases,fill=Bases))+geom_bar(stat = "identity")+theme_minimal()
```

6. Extract the sequence from chromosome 20 at position 1,000,000 to 1,000,020 and retrieve the complement sequence

```{r installBS2ss3,  echo=toMessage}
mySeq <- BSgenome.Hsapiens.UCSC.hg19$chr20[1000000:1000020]
complement(mySeq)
```

7. Write this complement sequence to a FASTA file.

```{r installBS2ssaa3ee,  echo=toMessage}
mySet <- DNAStringSet(complement(mySeq))
names(mySet) <- "testFasta"
writeXStringSet(mySet,filepath ="data/testFasta.fa")
```


8. Look up the position of MYC in IGV (Human hg19) and find the genomic coordinates of its first exon.

```{r installBS2ss3ssee,  echo=toMessage}
chr8:128748315-128748869
```

9. Extract the sequence for the first exon.

```{r installBS2sssss3ee,  echo=toMessage}
mySeq <- BSgenome.Hsapiens.UCSC.hg19$chr8[128748315:128748869]

```

10. Compare the sequence to that found in IGV and identify start of translated region in gene 


11. Count the number of classical start codons (ATG) in the first exon.

```{r installBS2ss3aaee,  echo=toMessage}
countPattern("ATG",mySeq)
```

12. Use IGV to review translation start codon for Gapdh and similarly count occurrence of ATG in exon 2 of NM_001289745 transcript. (chr12:6643976-6644027)

```{r installBS2ss3saasee,  echo=toMessage}
gapdhSeq <- BSgenome.Hsapiens.UCSC.hg19$chr12[6643976:6644027]
countPattern("ATG",gapdhSeq)

```

